{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dandreas/.conda/envs/deepV_a100/lib/python3.10/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "from saveAndLoad import *\n",
    "\n",
    "def printAllHead(df,n=5):\n",
    "    with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.max_colwidth', None):\n",
    "        print(df.head(n))\n",
    "def printAll(x):\n",
    "    with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.max_colwidth', None):\n",
    "        print(x)\n",
    "\n",
    "def inspect(df,c):\n",
    "    # print(df.columns)\n",
    "    print('nunique',df[c].nunique())\n",
    "    print(df[c].unique())\n",
    "    print(df[c].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2239204/3534989332.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  maf = pd.read_csv('mc3.v0.2.8.PUBLIC.maf',sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# https://gdc.cancer.gov/about-data/publications/pancanatlas\n",
    "maf = pd.read_csv('mc3.v0.2.8.PUBLIC.maf',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hugo_Symbol\n",
      "Entrez_Gene_Id\n",
      "Center\n",
      "NCBI_Build\n",
      "Chromosome\n",
      "Start_Position\n",
      "End_Position\n",
      "Strand\n",
      "Variant_Classification\n",
      "Variant_Type\n",
      "Reference_Allele\n",
      "Tumor_Seq_Allele1\n",
      "Tumor_Seq_Allele2\n",
      "dbSNP_RS\n",
      "dbSNP_Val_Status\n",
      "Tumor_Sample_Barcode\n",
      "Matched_Norm_Sample_Barcode\n",
      "Match_Norm_Seq_Allele1\n",
      "Match_Norm_Seq_Allele2\n",
      "Tumor_Validation_Allele1\n",
      "Tumor_Validation_Allele2\n",
      "Match_Norm_Validation_Allele1\n",
      "Match_Norm_Validation_Allele2\n",
      "Verification_Status\n",
      "Validation_Status\n",
      "Mutation_Status\n",
      "Sequencing_Phase\n",
      "Sequence_Source\n",
      "Validation_Method\n",
      "Score\n",
      "BAM_File\n",
      "Sequencer\n",
      "Tumor_Sample_UUID\n",
      "Matched_Norm_Sample_UUID\n",
      "HGVSc\n",
      "HGVSp\n",
      "HGVSp_Short\n",
      "Transcript_ID\n",
      "Exon_Number\n",
      "t_depth\n",
      "t_ref_count\n",
      "t_alt_count\n",
      "n_depth\n",
      "n_ref_count\n",
      "n_alt_count\n",
      "all_effects\n",
      "Allele\n",
      "Gene\n",
      "Feature\n",
      "Feature_type\n",
      "Consequence\n",
      "cDNA_position\n",
      "CDS_position\n",
      "Protein_position\n",
      "Amino_acids\n",
      "Codons\n",
      "Existing_variation\n",
      "ALLELE_NUM\n",
      "DISTANCE\n",
      "STRAND\n",
      "SYMBOL\n",
      "SYMBOL_SOURCE\n",
      "HGNC_ID\n",
      "BIOTYPE\n",
      "CANONICAL\n",
      "CCDS\n",
      "ENSP\n",
      "SWISSPROT\n",
      "TREMBL\n",
      "UNIPARC\n",
      "RefSeq\n",
      "SIFT\n",
      "PolyPhen\n",
      "EXON\n",
      "INTRON\n",
      "DOMAINS\n",
      "GMAF\n",
      "AFR_MAF\n",
      "AMR_MAF\n",
      "ASN_MAF\n",
      "EAS_MAF\n",
      "EUR_MAF\n",
      "SAS_MAF\n",
      "AA_MAF\n",
      "EA_MAF\n",
      "CLIN_SIG\n",
      "SOMATIC\n",
      "PUBMED\n",
      "MOTIF_NAME\n",
      "MOTIF_POS\n",
      "HIGH_INF_POS\n",
      "MOTIF_SCORE_CHANGE\n",
      "IMPACT\n",
      "PICK\n",
      "VARIANT_CLASS\n",
      "TSL\n",
      "HGVS_OFFSET\n",
      "PHENO\n",
      "MINIMISED\n",
      "ExAC_AF\n",
      "ExAC_AF_AFR\n",
      "ExAC_AF_AMR\n",
      "ExAC_AF_EAS\n",
      "ExAC_AF_FIN\n",
      "ExAC_AF_NFE\n",
      "ExAC_AF_OTH\n",
      "ExAC_AF_SAS\n",
      "GENE_PHENO\n",
      "FILTER\n",
      "COSMIC\n",
      "CENTERS\n",
      "CONTEXT\n",
      "DBVS\n",
      "NCALLERS\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>ExAC_AF_NFE</th>\n",
       "      <th>ExAC_AF_OTH</th>\n",
       "      <th>ExAC_AF_SAS</th>\n",
       "      <th>GENE_PHENO</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>CENTERS</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>DBVS</th>\n",
       "      <th>NCALLERS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TACC2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>123810032</td>\n",
       "      <td>123810032</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.T38M|c.113C&gt;T|3</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>GGACACGCCCG</td>\n",
       "      <td>by1000G</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JAKMIP3</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>10</td>\n",
       "      <td>133967449</td>\n",
       "      <td>133967449</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>CTGGACGAGGA</td>\n",
       "      <td>byFrequency</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PANX3</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>124489539</td>\n",
       "      <td>124489539</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.R296Q|c.887G&gt;A|3</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>ATGTCGGTGGG</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPI1</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>47380512</td>\n",
       "      <td>47380512</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>RADIA|MUSE</td>\n",
       "      <td>GGCTGGGGACA</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>11</td>\n",
       "      <td>89868837</td>\n",
       "      <td>89868837</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SITE|p.R65C|c.193C&gt;T|4</td>\n",
       "      <td>MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS</td>\n",
       "      <td>TTCTTCGGTAA</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600958</th>\n",
       "      <td>THOC2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>122829875</td>\n",
       "      <td>122829875</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>RADIA|MUSE|VARSCANS</td>\n",
       "      <td>CCTATTAAAGA</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600959</th>\n",
       "      <td>SLC25A14</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>129498779</td>\n",
       "      <td>129498779</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>SOMATICSNIPER|RADIA|MUTECT|MUSE|VARSCANS</td>\n",
       "      <td>AAAAAGCCTCT</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600960</th>\n",
       "      <td>FGF13</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>137714160</td>\n",
       "      <td>137714160</td>\n",
       "      <td>+</td>\n",
       "      <td>3'Flank</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>MUTECT|MUSE</td>\n",
       "      <td>AGTTCCTGTCT</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600961</th>\n",
       "      <td>XIST</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>73044379</td>\n",
       "      <td>73044379</td>\n",
       "      <td>+</td>\n",
       "      <td>RNA</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>RADIA|MUTECT|MUSE|VARSCANS</td>\n",
       "      <td>TTAAACCCTTG</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3600962</th>\n",
       "      <td>SRPX2</td>\n",
       "      <td>0</td>\n",
       "      <td>.</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>X</td>\n",
       "      <td>99925883</td>\n",
       "      <td>99925883</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NONE</td>\n",
       "      <td>RADIA|MUTECT|MUSE|VARSCANS</td>\n",
       "      <td>ACATGGAACCT</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3600963 rows Ã— 114 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  \\\n",
       "0             TACC2               0      .     GRCh37         10   \n",
       "1           JAKMIP3               0      .     GRCh37         10   \n",
       "2             PANX3               0      .     GRCh37         11   \n",
       "3              SPI1               0      .     GRCh37         11   \n",
       "4           NAALAD2               0      .     GRCh37         11   \n",
       "...             ...             ...    ...        ...        ...   \n",
       "3600958       THOC2               0      .     GRCh37          X   \n",
       "3600959    SLC25A14               0      .     GRCh37          X   \n",
       "3600960       FGF13               0      .     GRCh37          X   \n",
       "3600961        XIST               0      .     GRCh37          X   \n",
       "3600962       SRPX2               0      .     GRCh37          X   \n",
       "\n",
       "         Start_Position  End_Position Strand Variant_Classification  \\\n",
       "0             123810032     123810032      +      Missense_Mutation   \n",
       "1             133967449     133967449      +                 Silent   \n",
       "2             124489539     124489539      +      Missense_Mutation   \n",
       "3              47380512      47380512      +      Missense_Mutation   \n",
       "4              89868837      89868837      +      Missense_Mutation   \n",
       "...                 ...           ...    ...                    ...   \n",
       "3600958       122829875     122829875      +      Missense_Mutation   \n",
       "3600959       129498779     129498779      +                 Intron   \n",
       "3600960       137714160     137714160      +                3'Flank   \n",
       "3600961        73044379      73044379      +                    RNA   \n",
       "3600962        99925883      99925883      +      Missense_Mutation   \n",
       "\n",
       "        Variant_Type  ... ExAC_AF_NFE ExAC_AF_OTH ExAC_AF_SAS GENE_PHENO  \\\n",
       "0                SNP  ...           .           .           .          .   \n",
       "1                SNP  ...           .           .           .          .   \n",
       "2                SNP  ...           .           .           .          .   \n",
       "3                SNP  ...           .           .           .          .   \n",
       "4                SNP  ...           .           .           .          .   \n",
       "...              ...  ...         ...         ...         ...        ...   \n",
       "3600958          SNP  ...           .           .           .          .   \n",
       "3600959          SNP  ...           .           .           .          .   \n",
       "3600960          SNP  ...           .           .           .          .   \n",
       "3600961          SNP  ...           .           .           .          .   \n",
       "3600962          SNP  ...           .           .           .          .   \n",
       "\n",
       "        FILTER                   COSMIC  \\\n",
       "0         PASS   SITE|p.T38M|c.113C>T|3   \n",
       "1         PASS                     NONE   \n",
       "2         PASS  SITE|p.R296Q|c.887G>A|3   \n",
       "3         PASS                     NONE   \n",
       "4         PASS   SITE|p.R65C|c.193C>T|4   \n",
       "...        ...                      ...   \n",
       "3600958   PASS                     NONE   \n",
       "3600959   PASS                     NONE   \n",
       "3600960   PASS                     NONE   \n",
       "3600961   PASS                     NONE   \n",
       "3600962   PASS                     NONE   \n",
       "\n",
       "                                          CENTERS      CONTEXT         DBVS  \\\n",
       "0        MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS  GGACACGCCCG      by1000G   \n",
       "1        MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS  CTGGACGAGGA  byFrequency   \n",
       "2        MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS  ATGTCGGTGGG            .   \n",
       "3                                      RADIA|MUSE  GGCTGGGGACA            .   \n",
       "4        MUTECT|RADIA|SOMATICSNIPER|MUSE|VARSCANS  TTCTTCGGTAA            .   \n",
       "...                                           ...          ...          ...   \n",
       "3600958                       RADIA|MUSE|VARSCANS  CCTATTAAAGA            .   \n",
       "3600959  SOMATICSNIPER|RADIA|MUTECT|MUSE|VARSCANS  AAAAAGCCTCT            .   \n",
       "3600960                               MUTECT|MUSE  AGTTCCTGTCT            .   \n",
       "3600961                RADIA|MUTECT|MUSE|VARSCANS  TTAAACCCTTG            .   \n",
       "3600962                RADIA|MUTECT|MUSE|VARSCANS  ACATGGAACCT            .   \n",
       "\n",
       "        NCALLERS  \n",
       "0              5  \n",
       "1              5  \n",
       "2              5  \n",
       "3              2  \n",
       "4              5  \n",
       "...          ...  \n",
       "3600958        3  \n",
       "3600959        5  \n",
       "3600960        2  \n",
       "3600961        4  \n",
       "3600962        4  \n",
       "\n",
       "[3600963 rows x 114 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for i in maf.columns: print(i)\n",
    "maf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nunique 10295\n",
      "['TCGA-02-0003-01A-01D-1490-08' 'TCGA-02-0033-01A-01D-1490-08'\n",
      " 'TCGA-02-0047-01A-01D-1490-08' ... 'TCGA-ZT-A8OM-01A-11D-A428-09'\n",
      " 'TCGA-ZU-A8S4-01A-11D-A417-09' 'TCGA-ZX-AA5X-01A-11D-A42O-09']\n",
      "Tumor_Sample_Barcode\n",
      "TCGA-FW-A3R5-06A-11D-A23B-08    55283\n",
      "TCGA-A5-A0G2-01A-11W-A062-09    40047\n",
      "TCGA-EO-A22U-01A-11D-A17W-09    25776\n",
      "TCGA-FI-A2D5-01A-11D-A17D-09    24377\n",
      "TCGA-AX-A2HC-01A-11D-A17D-09    23246\n",
      "                                ...  \n",
      "TCGA-FB-A4P6-01A-12D-A26I-08        1\n",
      "TCGA-FP-8209-01A-11D-2340-08        1\n",
      "TCGA-HZ-7924-01A-11D-2154-08        1\n",
      "TCGA-AB-2825-03D-01W-0755-09        1\n",
      "TCGA-G9-6370-01A-11D-1786-08        1\n",
      "Name: count, Length: 10295, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "inspect(maf,'Tumor_Sample_Barcode')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2012970/3845924749.py:1: DtypeWarning: Columns (8,9,27,42,44,45,46,47,48,49,50,51,52,53,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,280,281,282,283,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,333,334,335,336,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,375,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,404,405,406,407,408,409,410,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,432,433,434,435,436,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,474,475,476,477,478,479,480,481,482,483,484,485,486,487,488,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,553,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,575,576,577,578,579,580,581,582,583,584,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600,601,602,603,604,605,606,607,608,609,610,611,612,613,614,615,616,617,618,619,620,621,622,623,624,625,626,627,628,629,630,631,632,633,634,635,636,637,639,640,641,642,643,644,645,646,647,648,649,650,651,652,653,654,655,656,657,658,659,660,661,662,663,664,665,666,667,668,669,670,671,672,673,674,675,676,677,678,679,680,681,682,683,684,685,686,687,688,689,690,691,692,693,694,695,696,697,698,699,700,701,702,703,704,705,706,707,708,709,710,711,712,713,714,715,716,717,718,719,720,721,722,723,724,725,726,727,728,729,730,731,732,733,734,735,736,737,738,739,740,741,742,743,744) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  fu = pd.read_csv('clinical_PANCAN_patient_with_followup.tsv',sep='\\t', encoding='latin1')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bcr_patient_uuid \t\t\t 10956\n",
      "informed_consent_verified \t\t\t 10956\n",
      "patient_id \t\t\t 10956\n",
      "bcr_patient_barcode \t\t\t 10956\n",
      "gender \t\t\t 10956\n",
      "acronym \t\t\t 10956\n",
      "history_of_neoadjuvant_treatment \t\t\t 10954\n",
      "icd_10 \t\t\t 10953\n",
      "icd_o_3_site \t\t\t 10953\n",
      "icd_o_3_histology \t\t\t 10953\n",
      "vital_status \t\t\t 10952\n",
      "tissue_source_site \t\t\t 10949\n",
      "age_at_initial_pathologic_diagnosis \t\t\t 10906\n",
      "tumor_tissue_site \t\t\t 10856\n",
      "days_to_birth \t\t\t 10847\n",
      "days_to_initial_pathologic_diagnosis \t\t\t 10832\n",
      "form_completion_date \t\t\t 10830\n",
      "year_of_initial_pathologic_diagnosis \t\t\t 10827\n",
      "histological_type \t\t\t 10391\n",
      "person_neoplasm_cancer_status \t\t\t 10270\n",
      "radiation_therapy \t\t\t 10008\n",
      "race \t\t\t 9919\n",
      "prior_dx \t\t\t 9815\n",
      "tissue_retrospective_collection_indicator \t\t\t 9723\n",
      "tissue_prospective_collection_indicator \t\t\t 9716\n",
      "ethnicity \t\t\t 9120\n",
      "days_to_last_followup \t\t\t 9091\n",
      "new_tumor_event_after_initial_treatment \t\t\t 8971\n",
      "pathologic_T \t\t\t 7886\n",
      "pathologic_N \t\t\t 7837\n",
      "system_version \t\t\t 7261\n",
      "pathologic_stage \t\t\t 7011\n",
      "pathologic_M \t\t\t 6968\n",
      "anatomic_neoplasm_subdivision \t\t\t 4908\n",
      "primary_lymph_node_presentation_assessment \t\t\t 4870\n",
      "initial_pathologic_diagnosis_method \t\t\t 4675\n",
      "neoplasm_histologic_grade \t\t\t 4637\n",
      "residual_tumor \t\t\t 4514\n",
      "number_of_lymphnodes_positive_by_he \t\t\t 3955\n",
      "lymph_node_examined_count \t\t\t 3870\n",
      "days_to_death \t\t\t 3470\n",
      "tobacco_smoking_history \t\t\t 3128\n",
      "postoperative_rx_tx \t\t\t 3079\n",
      "weight \t\t\t 2805\n",
      "primary_therapy_outcome_success \t\t\t 2772\n",
      "laterality \t\t\t 2754\n",
      "height \t\t\t 2730\n",
      "clinical_stage \t\t\t 2515\n",
      "eastern_cancer_oncology_group \t\t\t 2346\n",
      "targeted_molecular_therapy \t\t\t 2050\n",
      "karnofsky_performance_score \t\t\t 1925\n",
      "performance_status_scale_timing \t\t\t 1902\n",
      "menopause_status \t\t\t 1871\n",
      "clinical_M \t\t\t 1737\n",
      "margin_status \t\t\t 1724\n",
      "number_pack_years_smoked \t\t\t 1672\n",
      "clinical_T \t\t\t 1648\n",
      "kras_gene_analysis_performed \t\t\t 1272\n",
      "stopped_smoking_year \t\t\t 1183\n",
      "breast_carcinoma_progesterone_receptor_status \t\t\t 1099\n",
      "breast_carcinoma_estrogen_receptor_status \t\t\t 1099\n",
      "lab_proc_her2_neu_immunohistochemistry_receptor_status \t\t\t 1091\n",
      "additional_pharmaceutical_therapy \t\t\t 1087\n",
      "additional_radiation_therapy \t\t\t 1085\n",
      "breast_carcinoma_surgical_procedure_name \t\t\t 1046\n",
      "diagnosis \t\t\t 1026\n",
      "days_to_new_tumor_event_after_initial_treatment \t\t\t 1022\n",
      "lab_procedure_her2_neu_in_situ_hybrid_outcome_type \t\t\t 1017\n",
      "year_of_tobacco_smoking_onset \t\t\t 1014\n",
      "clinical_N \t\t\t 1000\n",
      "axillary_lymph_node_stage_method_type \t\t\t 884\n",
      "alcohol_history_documented \t\t\t 866\n",
      "number_of_lymphnodes_positive_by_ihc \t\t\t 835\n",
      "lymphovascular_invasion_present \t\t\t 813\n",
      "lymphatic_invasion \t\t\t 796\n",
      "hemoglobin_result \t\t\t 769\n",
      "platelet_qualitative_result \t\t\t 759\n",
      "white_cell_count_result \t\t\t 755\n",
      "cytokeratin_immunohistochemistry_staining_method_micrometastasis_indicator \t\t\t 750\n",
      "venous_invasion \t\t\t 716\n",
      "serum_calcium_result \t\t\t 673\n",
      "perineural_invasion_present \t\t\t 647\n",
      "pulmonary_function_test_performed \t\t\t 642\n",
      "egfr_mutation_performed \t\t\t 631\n",
      "her2_immunohistochemistry_level_result \t\t\t 623\n",
      "prior_glioma \t\t\t 596\n",
      "family_history_of_cancer \t\t\t 590\n",
      "eml4_alk_translocation_performed \t\t\t 589\n",
      "total_pelv_lnr \t\t\t 584\n",
      "surgical_approach \t\t\t 575\n",
      "total_aor_lnr \t\t\t 568\n",
      "braf_gene_analysis_performed \t\t\t 567\n",
      "location_in_lung_parenchyma \t\t\t 566\n",
      "synchronous_colon_cancer_present \t\t\t 564\n",
      "peritoneal_wash \t\t\t 560\n",
      "history_of_colon_polyps \t\t\t 548\n",
      "city_of_procurement \t\t\t 546\n",
      "number_of_first_degree_relatives_with_cancer_diagnosis \t\t\t 540\n",
      "new_neoplasm_event_type \t\t\t 535\n",
      "lymphnode_neck_dissection \t\t\t 525\n",
      "tumor_residual_disease \t\t\t 520\n",
      "pct_tumor_invasion \t\t\t 520\n",
      "tumor_location \t\t\t 515\n",
      "reflux_history \t\t\t 510\n",
      "genotyping_results_gene_mutation_not_reported_reason \t\t\t 507\n",
      "primary_thyroid_gland_neoplasm_location_anatomic_site \t\t\t 501\n",
      "zone_of_origin \t\t\t 500\n",
      "gleason_score \t\t\t 500\n",
      "secondary_pattern \t\t\t 500\n",
      "primary_pattern \t\t\t 500\n",
      "primary_neoplasm_focus_type \t\t\t 497\n",
      "history_ionizing_rt_to_head \t\t\t 496\n",
      "lymphnodes_examined \t\t\t 495\n",
      "seizure_history \t\t\t 489\n",
      "supratentorial_localization \t\t\t 486\n",
      "lymph_node_preoperative_scan_indicator \t\t\t 484\n",
      "total_pelv_lnp \t\t\t 484\n",
      "loss_expression_of_mismatch_repair_proteins_by_ihc \t\t\t 481\n",
      "neoplasm_length \t\t\t 478\n",
      "headache_history \t\t\t 473\n",
      "mental_status_changes \t\t\t 472\n",
      "primary_neoplasm_melanoma_dx \t\t\t 471\n",
      "er_level_cell_percentage_category \t\t\t 470\n",
      "tumor_tissue_site.1 \t\t\t 469\n",
      "visual_changes \t\t\t 468\n",
      "motor_movement_changes \t\t\t 467\n",
      "sensory_changes \t\t\t 466\n",
      "days_to_submitted_specimen_dx \t\t\t 465\n",
      "barretts_esophagus \t\t\t 462\n",
      "first_presenting_symptom \t\t\t 462\n",
      "person_lifetime_risk_radiation_exposure_indicator \t\t\t 461\n",
      "patient_personal_medical_history_thyroid_gland_disorder_name \t\t\t 456\n",
      "ldh1_mutation_tested \t\t\t 454\n",
      "days_to_psa \t\t\t 449\n",
      "first_presenting_symptom_longest_duration \t\t\t 444\n",
      "psa_value \t\t\t 443\n",
      "pln_pos_light_micro \t\t\t 441\n",
      "melanoma_origin_skin_anatomic_site \t\t\t 436\n",
      "preoperative_corticosteroids \t\t\t 436\n",
      "family_history_of_primary_brain_tumor \t\t\t 433\n",
      "biochemical_recurrence \t\t\t 431\n",
      "h_pylori_infection \t\t\t 430\n",
      "progesterone_receptor_level_cell_percent_category \t\t\t 430\n",
      "neoplasm_width \t\t\t 429\n",
      "preoperative_antiseizure_meds \t\t\t 427\n",
      "number_of_lymphnodes_examined \t\t\t 425\n",
      "pre_bronchodilator_fev1_percent \t\t\t 425\n",
      "tumor_type \t\t\t 420\n",
      "primary_tumor_multiple_present_ind \t\t\t 417\n",
      "asthma_history \t\t\t 412\n",
      "specimen_collection_method_name \t\t\t 410\n",
      "eczema_history \t\t\t 410\n",
      "diagnosis_subtype \t\t\t 408\n",
      "neoplasm_depth \t\t\t 407\n",
      "total_aor-lnp \t\t\t 406\n",
      "diagnostic_ct_abd_pelvis_performed \t\t\t 405\n",
      "mold_or_dust_allergy_history \t\t\t 403\n",
      "hay_fever_history \t\t\t 403\n",
      "preoperative_pretreatment_cea_level \t\t\t 402\n",
      "diagnostic_mri_performed \t\t\t 398\n",
      "lactate_dehydrogenase_result \t\t\t 397\n",
      "distant_metastasis_present_ind2 \t\t\t 395\n",
      "animal_insect_allergy_history \t\t\t 386\n",
      "food_allergy_history \t\t\t 382\n",
      "lymph_node_preoperative_assessment_diagnostic_imaging_type \t\t\t 375\n",
      "days_to_definitive_surgical_procedure_performed \t\t\t 374\n",
      "frequency_of_alcohol_consumption \t\t\t 369\n",
      "family_history_of_stomach_cancer \t\t\t 369\n",
      "relative_family_cancer_history_ind_3 \t\t\t 366\n",
      "melanoma_ulceration_indicator \t\t\t 365\n",
      "melanoma_clark_level_value \t\t\t 365\n",
      "presence_of_pathological_nodal_extracapsular_spread \t\t\t 363\n",
      "dlco_predictive_percent \t\t\t 360\n",
      "pre_bronchodilator_fev1_fvc_percent \t\t\t 360\n",
      "breslow_depth_value \t\t\t 360\n",
      "aln_pos_light_micro \t\t\t 358\n",
      "child_pugh_classification_grade \t\t\t 356\n",
      "tumor_level \t\t\t 356\n",
      "person_concomitant_prostate_carcinoma_occurrence_indicator \t\t\t 348\n",
      "age_began_smoking_in_years \t\t\t 339\n",
      "inherited_genetic_syndrome_found \t\t\t 332\n",
      "genotype_analysis_performed_indicator \t\t\t 329\n",
      "amount_of_alcohol_consumption_per_day \t\t\t 315\n",
      "birth_control_pill_history_usage_category \t\t\t 314\n",
      "state_province_country_of_procurement \t\t\t 313\n",
      "lymphnode_dissection_method_right \t\t\t 312\n",
      "lab_procedure_platelet_result_specified_value \t\t\t 310\n",
      "laboratory_prcoedure_platelet_result_lower_limit_of_normal_value \t\t\t 309\n",
      "laboratory_prcoedure_platelet_result_upper_limit_of_normal_value \t\t\t 309\n",
      "hist_of_non_mibc \t\t\t 307\n",
      "laboratory_procedure_creatinine_result_upper_limit_of_normal_value \t\t\t 307\n",
      "laboratory_procedure_total_bilirubin_result_specified_upper_limit_of_normal_value \t\t\t 306\n",
      "laboratory_procedure_creatinine_result_lower_limit_of_normal_value \t\t\t 305\n",
      "hematology_serum_creatinine_laboratory_result_value_in_mg_dl \t\t\t 305\n",
      "laboratory_procedure_albumin_result_specified_value \t\t\t 303\n",
      "colon_polyps_present \t\t\t 303\n",
      "laboratory_procedure_total_bilirubin_result_upper_limit_normal_value \t\t\t 302\n",
      "non_nodal_tumor_deposits \t\t\t 302\n",
      "laboratory_procedure_total_bilirubin_result_specified_lower_limit_of_normal_value \t\t\t 301\n",
      "laboratory_procedure_prothrombin_time_result_value \t\t\t 300\n",
      "laboratory_procedure_albumin_result_upper_limit_of_normal_value \t\t\t 298\n",
      "laboratory_procedure_albumin_result_lower_limit_of_normal_value \t\t\t 298\n",
      "viral_hepatitis_serology \t\t\t 296\n",
      "person_occupation_description_text \t\t\t 293\n",
      "lymphnode_dissection_method_left \t\t\t 292\n",
      "laboratory_procedure_alpha_fetoprotein_outcome_value \t\t\t 284\n",
      "days_to_initial_score_performance_status_scale \t\t\t 283\n",
      "ablation_embolization_tx_adjuvant \t\t\t 282\n",
      "liver_fibrosis_ishak_score_category \t\t\t 278\n",
      "pln_pos_ihc \t\t\t 277\n",
      "history_hepato_carcinoma_risk_factor \t\t\t 276\n",
      "surgical_procedure_purpose_other_text \t\t\t 276\n",
      "days_to_diagnostic_computed_tomography_performed \t\t\t 272\n",
      "laboratory_procedure_alpha_fetoprotein_outcome_lower_limit_of_normal_value \t\t\t 272\n",
      "laboratory_procedure_international_normalization_ratio_result_upper_limit_of_normal_value \t\t\t 271\n",
      "antireflux_treatment \t\t\t 271\n",
      "laboratory_procedure_international_normalization_ratio_result_lower_limit_of_normal_value \t\t\t 271\n",
      "laboratory_procedure_alpha_fetoprotein_outcome_upper_limit_of_normal_value \t\t\t 270\n",
      "new_neoplasm_event_occurrence_anatomic_site \t\t\t 269\n",
      "total_number_of_pregnancies \t\t\t 268\n",
      "number_of_successful_pregnancies_which_resulted_in_at_least_1_live_birth \t\t\t 263\n",
      "hysterectomy_performed_type \t\t\t 262\n",
      "occupation_primary_job \t\t\t 261\n",
      "ct_scan \t\t\t 260\n",
      "bone_scan_results \t\t\t 249\n",
      "erythrocyte_sedimentation_rate_result \t\t\t 247\n",
      "days_to_bone_scan_performed \t\t\t 243\n",
      "positive_finding_estrogen_receptor_other_measurement_scale_text \t\t\t 243\n",
      "country_of_procurement \t\t\t 242\n",
      "aln_pos_ihc \t\t\t 237\n",
      "init_pathology_dx_method_other \t\t\t 237\n",
      "hpv_status_by_p16_testing \t\t\t 237\n",
      "post_bronchodilator_fev1_percent \t\t\t 235\n",
      "post_surgical_procedure_assessment_thyroid_gland_carcinoma_status \t\t\t 234\n",
      "her2_neu_chromosone_17_signal_ratio_value \t\t\t 233\n",
      "number_of_lymphnodes_positive \t\t\t 229\n",
      "hpv_status_by_ish_testing \t\t\t 227\n",
      "tumor_depth \t\t\t 226\n",
      "pos_finding_progesterone_receptor_other_measurement_scale_text \t\t\t 222\n",
      "er_detection_method_text \t\t\t 218\n",
      "pathologic_tumor_length \t\t\t 216\n",
      "new_tumor_dx_prior_submitted_specimen_dx \t\t\t 215\n",
      "immunohistochemistry_positive_cell_score \t\t\t 214\n",
      "tumor_multifocal \t\t\t 213\n",
      "pgr_detection_method_text \t\t\t 211\n",
      "disease_extracapsular_extension_ind-3 \t\t\t 210\n",
      "her2_erbb_pos_finding_cell_percent_category \t\t\t 208\n",
      "maximum_tumor_dimension \t\t\t 208\n",
      "breast_carcinoma_immunohistochemistry_pos_cell_score \t\t\t 208\n",
      "pathologic_tumor_width \t\t\t 207\n",
      "pathologic_tumor_depth \t\t\t 203\n",
      "diagnostic_ct_abd_pelvis_result \t\t\t 203\n",
      "radiation_therapy.1 \t\t\t 202\n",
      "postoperative_rx_tx.1 \t\t\t 202\n",
      "post_bronchodilator_fev1_fvc_percent \t\t\t 198\n",
      "primary_anatomic_site_count \t\t\t 194\n",
      "i_131_total_administered_preparation_technique \t\t\t 193\n",
      "patient_death_reason \t\t\t 190\n",
      "person_primary_industry_text \t\t\t 186\n",
      "metastatic_neoplasm_confirmed \t\t\t 186\n",
      "adenocarcinoma_invasion \t\t\t 185\n",
      "esophageal_tumor_cental_location \t\t\t 184\n",
      "egfr_amplication_status \t\t\t 184\n",
      "esophageal_tumor_involvement_site \t\t\t 184\n",
      "surgery_performed_type \t\t\t 183\n",
      "history_of_esophageal_cancer \t\t\t 183\n",
      "local_disease_recurrence \t\t\t 183\n",
      "tumor_total_necrosis_percent \t\t\t 182\n",
      "family_medical_history_relative_family_member_relationship_type \t\t\t 180\n",
      "history_pheo_or_para_include_benign \t\t\t 179\n",
      "keratinizing_squamous_cell_carcinoma_present_indicator \t\t\t 175\n",
      "history_of_diabetes \t\t\t 173\n",
      "primary_melanoma_at_diagnosis_count \t\t\t 173\n",
      "malignant_neoplasm_mitotic_count_rate \t\t\t 173\n",
      "adjacent_hepatic_tissue_inflammation_extent_type \t\t\t 171\n",
      "history_of_chronic_pancreatitis \t\t\t 168\n",
      "days_to_performance_status_assessment \t\t\t 164\n",
      "lymphovascular_invasion_indicator \t\t\t 162\n",
      "pregnant_at_diagnosis \t\t\t 160\n",
      "post_op_ablation_embolization_tx \t\t\t 160\n",
      "initial_diagnosis_by \t\t\t 160\n",
      "lymph_node_metastasis_radiographic_evidence \t\t\t 160\n",
      "p53_gene_analysis \t\t\t 158\n",
      "extrathyroid_carcinoma_present_extension_status \t\t\t 154\n",
      "tertiary_pattern \t\t\t 154\n",
      "cancer_diagnosis_cancer_type_icd9_text_name \t\t\t 150\n",
      "vascular_tumor_cell_invasion_type \t\t\t 148\n",
      "breast_carcinoma_immunohistochemistry_er_pos_finding_scale \t\t\t 147\n",
      "patient_pregnancy_spontaneous_abortion_count \t\t\t 147\n",
      "disease_detected_on_screening \t\t\t 146\n",
      "mitotic_count \t\t\t 143\n",
      "breast_carcinoma_immunohistochemistry_progesterone_receptor_pos_finding_scale \t\t\t 142\n",
      "kras_mutation_found \t\t\t 141\n",
      "columnar_metaplasia_present \t\t\t 137\n",
      "microsatellite_instability \t\t\t 135\n",
      "family_history_testicular_cancer \t\t\t 134\n",
      "histological_percentage \t\t\t 134\n",
      "history_of_undescended_testis \t\t\t 133\n",
      "history_fertility \t\t\t 133\n",
      "cervical_carcinoma_corpus_uteri_involvement_indicator \t\t\t 132\n",
      "serum_markers \t\t\t 132\n",
      "family_history_other_cancer \t\t\t 132\n",
      "radiologic_tumor_length \t\t\t 131\n",
      "columnar_mucosa_dysplasia \t\t\t 130\n",
      "history_hypospadias \t\t\t 130\n",
      "intratubular_germ_cell_neoplasm \t\t\t 129\n",
      "ldh1_mutation_found \t\t\t 129\n",
      "radiologic_tumor_width \t\t\t 126\n",
      "columnar_mucosa_goblet_cell_present \t\t\t 126\n",
      "post_orchi_lymph_node_dissection \t\t\t 125\n",
      "days_to_pre_orchi_serum_test \t\t\t 125\n",
      "ldh1_mutation_test_method \t\t\t 124\n",
      "first_treatment_success \t\t\t 124\n",
      "patient_pregnancy_therapeutic_abortion_count \t\t\t 123\n",
      "circumferential_resection_margin \t\t\t 123\n",
      "pre_orchi_afp \t\t\t 123\n",
      "history_myasthenia_gravis \t\t\t 123\n",
      "masaoka_stage \t\t\t 122\n",
      "state_province_of_procurement \t\t\t 121\n",
      "testis_tumor_macroextent \t\t\t 119\n",
      "days_to_post_orchi_serum_test \t\t\t 118\n",
      "cancer_diagnosis_first_degree_relative_number \t\t\t 118\n",
      "ectopic_pregnancy_count \t\t\t 116\n",
      "new_neoplasm_occurrence_anatomic_site_text \t\t\t 115\n",
      "molecular_test_result \t\t\t 115\n",
      "post_orchi_afp \t\t\t 114\n",
      "her2_neu_breast_carcinoma_copy_analysis_input_total_number \t\t\t 114\n",
      "pre_orchi_ldh \t\t\t 114\n",
      "radiation_therapy_administered_preparation_technique_text \t\t\t 113\n",
      "pregnancy_stillbirth_count \t\t\t 112\n",
      "histological_type_other \t\t\t 110\n",
      "post_orchi_ldh \t\t\t 110\n",
      "antireflux_treatment_type \t\t\t 108\n",
      "diabetes \t\t\t 107\n",
      "hypertension \t\t\t 107\n",
      "pregnancies \t\t\t 106\n",
      "fluorescence_in_situ_hybridization_diagnostic_procedure_chromosome_17_signal_result_range \t\t\t 106\n",
      "breast_carcinoma_primary_surgical_procedure_name \t\t\t 106\n",
      "histologic_grading_tier_category \t\t\t 105\n",
      "primary_tumor_lower_uterus_segment \t\t\t 105\n",
      "colorectal_cancer \t\t\t 104\n",
      "her2_neu_and_centromere_17_copy_number_analysis_input_total_number_count \t\t\t 104\n",
      "leiomyosarcoma_major_vessel_involvement \t\t\t 104\n",
      "leiomyosarcoma_histologic_subtype \t\t\t 104\n",
      "country_of_birth \t\t\t 101\n",
      "metastatic_breast_carcinoma_lab_proc_her2_neu_immunohistochemistry_receptor_status \t\t\t 100\n",
      "radiologic_tumor_depth \t\t\t 99\n",
      "horm_ther \t\t\t 98\n",
      "metastatic_breast_carcinoma_lab_proc_her2_neu_in_situ_hybridization_outcome_type \t\t\t 98\n",
      "metastatic_breast_carcinoma_estrogen_receptor_status \t\t\t 98\n",
      "metastatic_breast_carcinoma_progesterone_receptor_status \t\t\t 98\n",
      "pre_orchi_hcg \t\t\t 97\n",
      "specific_tumor_total_necrosis_percent \t\t\t 96\n",
      "prior_tamoxifen_administered_usage_category \t\t\t 95\n",
      "days_to_diagnostic_mri_performed \t\t\t 93\n",
      "mitotane_therapy \t\t\t 91\n",
      "rt_administered_type \t\t\t 90\n",
      "her2_erbb_method_calculation_method_text \t\t\t 90\n",
      "presence_of_sarcomatoid_features \t\t\t 89\n",
      "molecular_analysis_performed_indicator \t\t\t 88\n",
      "necrosis \t\t\t 87\n",
      "pleurodesis_performed_prior \t\t\t 86\n",
      "patient_history_immune_system_and_related_disorders_name \t\t\t 86\n",
      "chemotherapy_regimen_type \t\t\t 85\n",
      "i_131_first_administered_dose \t\t\t 84\n",
      "rt_pelvis_administered_total_dose \t\t\t 83\n",
      "invasion_of_tumor_capsule \t\t\t 83\n",
      "alcoholic_exposure_category \t\t\t 83\n",
      "person_concomitant_prostate_carcinoma_pathologic_t_stage \t\t\t 83\n",
      "diagnostic_mri_result \t\t\t 82\n",
      "history_asbestos_exposure \t\t\t 82\n",
      "bladder_carcinoma_extracapsular_extension_status \t\t\t 81\n",
      "chemical_exposure_text \t\t\t 80\n",
      "tumor_thickness \t\t\t 80\n",
      "histological_type.1 \t\t\t 80\n",
      "tumor_thickness_measurement \t\t\t 80\n",
      "tumor_morphology_percentage \t\t\t 80\n",
      "mitotic_rate \t\t\t 80\n",
      "weiss_venous_invasion \t\t\t 79\n",
      "tumor_basal_diameter \t\t\t 79\n",
      "tumor_basal_diameter_mx \t\t\t 79\n",
      "ct_scan_findings \t\t\t 77\n",
      "cervical_neoplasm_pathologic_margin_involved_type \t\t\t 77\n",
      "malignant_neoplasm_metastatic_involvement_site \t\t\t 76\n",
      "distant_metastasis_anatomic_site \t\t\t 76\n",
      "number_of_relatives_with_stomach_cancer \t\t\t 76\n",
      "pathologic_tumor_burden \t\t\t 75\n",
      "extrascleral_extension \t\t\t 75\n",
      "dose_frequency_text \t\t\t 74\n",
      "number_of_loci_tested \t\t\t 74\n",
      "agent_total_dose_count \t\t\t 74\n",
      "i_131_total_administered_dose \t\t\t 73\n",
      "female_breast_feeding_or_pregnancy_status_indicator \t\t\t 72\n",
      "number_of_abnormal_loci \t\t\t 72\n",
      "source_of_patient_death_reason \t\t\t 72\n",
      "mesothelioma_detection_method \t\t\t 72\n",
      "planned_surgery_status \t\t\t 72\n",
      "gene_expression_profile \t\t\t 72\n",
      "nuclear_grade_III_IV \t\t\t 71\n",
      "eye_color \t\t\t 71\n",
      "weiss_score \t\t\t 70\n",
      "days_to_first_biochemical_recurrence \t\t\t 70\n",
      "diffuse_architecture \t\t\t 70\n",
      "brachytherapy_method_type \t\t\t 70\n",
      "cytoplasm_presence_less_than_equal_25_percent \t\t\t 69\n",
      "sinusoid_invasion \t\t\t 68\n",
      "atypical_mitotic_figures \t\t\t 68\n",
      "family_member_relationship_type \t\t\t 68\n",
      "cytogenetic_abnormality \t\t\t 67\n",
      "postoperative_tx \t\t\t 66\n",
      "post_orchi_hcg \t\t\t 66\n",
      "loss_expression_of_mismatch_repair_proteins_by_ihc_result \t\t\t 65\n",
      "relative_cancer_type \t\t\t 64\n",
      "tumor_infiltrating_lymphocytes \t\t\t 64\n",
      "diagnostic_ct_result_outcome \t\t\t 64\n",
      "tumor_shape_pathologic_clinical \t\t\t 63\n",
      "human_papillomavirus_type \t\t\t 62\n",
      "tumor_infiltrating_macrophages \t\t\t 62\n",
      "non_mibc_tx \t\t\t 61\n",
      "mitoses_count \t\t\t 61\n",
      "lymph_node_location_positive_pathology_name \t\t\t 61\n",
      "contiguous_organ_invaded \t\t\t 60\n",
      "primary_occupation \t\t\t 60\n",
      "tumor_tissue_site_other \t\t\t 60\n",
      "brachytherapy_first_reference_point_administered_total_dose \t\t\t 59\n",
      "extravascular_matrix_patterns \t\t\t 59\n",
      "breast_cancer_surgery_margin_status \t\t\t 59\n",
      "excess_adrenal_hormone_history_type \t\t\t 58\n",
      "metastatic_site_at_diagnosis \t\t\t 56\n",
      "treatment_prior_to_surgery \t\t\t 55\n",
      "excess_adrenal_hormone_diagnosis_method_type \t\t\t 54\n",
      "mitotane_therapy_adjuvant_setting \t\t\t 51\n",
      "her2_erbb_pos_finding_fluorescence_in_situ_hybridization_calculation_method_text \t\t\t 50\n",
      "primary_occupation_other \t\t\t 49\n",
      "tumor_response_cdus_type \t\t\t 49\n",
      "radiologic_tumor_burden \t\t\t 48\n",
      "hiv_positive_status \t\t\t 48\n",
      "discontiguous_lesion_count \t\t\t 48\n",
      "relative_family_cancer_hx_text \t\t\t 47\n",
      "first_degree_relative_history_thyroid_gland_carcinoma_diagnosis_relationship_type \t\t\t 46\n",
      "immunophenotypic_analysis_test \t\t\t 46\n",
      "result_of_immunophenotypic_analysis \t\t\t 46\n",
      "extranodal_involvement \t\t\t 46\n",
      "immunophenotypic_analysis_methodology \t\t\t 46\n",
      "creatinine_prior_tx \t\t\t 45\n",
      "creatinine_norm_range_upper \t\t\t 45\n",
      "creatinine_norm_range_lower \t\t\t 45\n",
      "chemotherapy_negation_radiation_therapy_concurrent_not_administered_reason \t\t\t 45\n",
      "ebv_antibody_status \t\t\t 44\n",
      "igcccg_stage \t\t\t 43\n",
      "nodal_anatomic_site \t\t\t 43\n",
      "asbestos_exposure_source \t\t\t 43\n",
      "well_differentiated_liposarcoma_primary_dx \t\t\t 43\n",
      "days_to_fdg_or_ct_pet_performed \t\t\t 43\n",
      "therapeutic_mitotane_levels_achieved \t\t\t 43\n",
      "interferon_90_day_prior_excision_admin_indicator \t\t\t 42\n",
      "number_of_relatives_diagnosed \t\t\t 41\n",
      "person_occupation_years_number \t\t\t 40\n",
      "asbestos_exposure_type \t\t\t 40\n",
      "percentage_of_follicular_component \t\t\t 40\n",
      "goblet_cells_present \t\t\t 39\n",
      "project_code \t\t\t 38\n",
      "disease_code \t\t\t 38\n",
      "number_of_involved_extranodal_sites \t\t\t 37\n",
      "mibc_90day_post_resection_bcg \t\t\t 37\n",
      "radiation_therapy_not_administered_reason \t\t\t 37\n",
      "induction_course_complete \t\t\t 37\n",
      "hist_hepato_carc_fact \t\t\t 36\n",
      "testis_tumor_microextent \t\t\t 36\n",
      "relative_family_cancer_history \t\t\t 36\n",
      "maint_therapy_course_complete \t\t\t 36\n",
      "prior_systemic_therapy_type \t\t\t 36\n",
      "first_diagnosis_age_asth_ecz_hay_fev_mold_dust \t\t\t 35\n",
      "complete_response_observed \t\t\t 35\n",
      "bone_marrow_involvement \t\t\t 35\n",
      "braf_gene_analysis_result \t\t\t 35\n",
      "concurrent_chemotherapy_dose \t\t\t 35\n",
      "cervical_neoplasm_pathologic_margin_involved_text \t\t\t 35\n",
      "platelet_result_lower_limit \t\t\t 34\n",
      "fdg_or_ct_pet_performed_outcome \t\t\t 34\n",
      "platelet_result_upper_limit \t\t\t 34\n",
      "platelet_result_count \t\t\t 34\n",
      "creatinine_value_in_mg_dl \t\t\t 34\n",
      "creatinine_upper_limit \t\t\t 34\n",
      "patient_personal_medical_history_thyroid_other_specify_text \t\t\t 34\n",
      "creatinine_lower_level \t\t\t 34\n",
      "history_hepato_carcinoma_risk_factor_other \t\t\t 34\n",
      "ldh_upper_limit \t\t\t 33\n",
      "prothrombin_time_result_value \t\t\t 32\n",
      "total_bilirubin_upper_limit \t\t\t 32\n",
      "bilirubin_upper_limit \t\t\t 32\n",
      "bilirubin_lower_limit \t\t\t 32\n",
      "ca_19_9_level_upper \t\t\t 31\n",
      "amt_alcohol_consumption_per_day \t\t\t 31\n",
      "albumin_result_specified_value \t\t\t 31\n",
      "albumin_result_lower_limit \t\t\t 31\n",
      "albumin_result_upper_limit \t\t\t 31\n",
      "ca_19_9_level \t\t\t 30\n",
      "ldh_lab_value \t\t\t 30\n",
      "anatomic_neoplasm_subdivision_other \t\t\t 30\n",
      "kras_mutation_codon \t\t\t 30\n",
      "metastatic_neoplasm_confirmed_diagnosis_method_name \t\t\t 30\n",
      "family_history_cancer_type_other \t\t\t 30\n",
      "bone_marrow_biopsy_done \t\t\t 30\n",
      "maximum_tumor_bulk_anatomic_location \t\t\t 30\n",
      "family_history_cancer_type \t\t\t 30\n",
      "eml4_alk_translocation_method \t\t\t 29\n",
      "fibrosis_ishak_score \t\t\t 28\n",
      "ca_19_9_level_lower \t\t\t 28\n",
      "braf_gene_genotyping_outcome_lab_results_text \t\t\t 27\n",
      "inter_norm_ratio_lower_limit \t\t\t 27\n",
      "intern_norm_ratio_upper_limit \t\t\t 27\n",
      "mib1_positive_percentage_range \t\t\t 27\n",
      "jewish_origin \t\t\t 26\n",
      "surgical_procedure_name_other_specify_text \t\t\t 26\n",
      "egfr_mutation_result \t\t\t 26\n",
      "fetoprotein_outcome_upper_limit \t\t\t 25\n",
      "extranodal_disease_involvement_site \t\t\t 25\n",
      "fetoprotein_outcome_lower_limit \t\t\t 25\n",
      "contiguous_organ_resection_site \t\t\t 25\n",
      "first_nonlymph_node_metastasis_anatomic_site \t\t\t 24\n",
      "fetoprotein_outcome_value \t\t\t 24\n",
      "undescended_testis_corrected \t\t\t 23\n",
      "level_of_non_descent \t\t\t 23\n",
      "igh_genotype_results \t\t\t 23\n",
      "kras_mutation_result \t\t\t 22\n",
      "suv_of_pleura_max \t\t\t 22\n",
      "igk_genotype_results \t\t\t 22\n",
      "family_cancer_type_txt \t\t\t 21\n",
      "therapeutic_mitotane_lvl_recurrence \t\t\t 21\n",
      "cancer_first_degree_relative \t\t\t 20\n",
      "outside_adrenal \t\t\t 20\n",
      "diagnosis_age \t\t\t 20\n",
      "axillary_lymph_node_stage_other_method_descriptive_text \t\t\t 20\n",
      "food_allergy_types \t\t\t 20\n",
      "metastatic_neoplasm_initial_diagnosis_anatomic_site \t\t\t 20\n",
      "abnormality_tested_results \t\t\t 19\n",
      "abnormality_tested_methodology \t\t\t 19\n",
      "genetic_abnormality_tested \t\t\t 19\n",
      "standardized_uptake_value_cervix_uteri_assessment_measurement \t\t\t 17\n",
      "undescended_testis_corrected_age \t\t\t 17\n",
      "history_pheo_or_para_anatomic_site \t\t\t 16\n",
      "asbestos_exposure_years \t\t\t 16\n",
      "radiosensitizing_agent_administered_indicator \t\t\t 16\n",
      "serum_mesothelin_lower_limit \t\t\t 16\n",
      "serum_mesothelin_upper_limit \t\t\t 16\n",
      "serum_mesothelin_prior_tx \t\t\t 15\n",
      "primary_occupation_years_worked \t\t\t 15\n",
      "prior_immunologic_disease_type \t\t\t 14\n",
      "pos_finding_her2_erbb2_other_measurement_scale_text \t\t\t 14\n",
      "mitotane_therapy_for_macroscopic_residual_disease \t\t\t 14\n",
      "breast_neoplasm_other_surgical_procedure_descriptive_text \t\t\t 14\n",
      "relation_testicular_cancer \t\t\t 14\n",
      "diagnostic_mri_result_outcome \t\t\t 14\n",
      "surgery_performed_type_other \t\t\t 14\n",
      "animal_insect_allergy_types \t\t\t 14\n",
      "days_to_diabetes_onset \t\t\t 14\n",
      "abnormality_tested_results.1 \t\t\t 13\n",
      "abnormality_tested_methodology.1 \t\t\t 13\n",
      "mpnst_neurofibromatosis \t\t\t 13\n",
      "other_genetic_abnormality_tested \t\t\t 13\n",
      "undescended_testis_method_left \t\t\t 13\n",
      "human_papillomavirus_laboratory_procedure_performed_name \t\t\t 12\n",
      "asbestos_exposure_age \t\t\t 12\n",
      "mpnst_exisiting_plexiform_neurofibroma \t\t\t 12\n",
      "metastatic_site_at_diagnosis_other \t\t\t 11\n",
      "days_to_pancreatitis_onset \t\t\t 11\n",
      "mpnst_nf1_genetic_testing \t\t\t 11\n",
      "cervical_carcinoma_pelvic_extension_text \t\t\t 11\n",
      "external_beam_radiation_therapy_administered_paraaortic_region_lymph_node_dose \t\t\t 11\n",
      "days_to_last_known_alive \t\t\t 11\n",
      "ras_family_gene_genotyping_outcome_lab_results_text \t\t\t 11\n",
      "ret_ptc_rearrangement_genotyping_outcome_lab_results_text \t\t\t 11\n",
      "human_papillomavirus_other_type_text \t\t\t 10\n",
      "first_diagnosis_age_of_animal_insect_allergy \t\t\t 10\n",
      "section_myasthenia_gravis \t\t\t 10\n",
      "therapeutic_mitotane_lvl_macroscopic_residual \t\t\t 10\n",
      "asbestos_exposure_age_last \t\t\t 9\n",
      "pleurodesis_performed_90_days \t\t\t 9\n",
      "therapeutic_mitotane_lvl_progression \t\t\t 8\n",
      "hysterectomy_performed_text \t\t\t 8\n",
      "other_contiguous_organ_resection_site \t\t\t 8\n",
      "other_metastatic_involvement_anatomic_site \t\t\t 8\n",
      "other_chemotherapy_agent_administration_specify \t\t\t 8\n",
      "ss18_ssx_testing_method \t\t\t 8\n",
      "oligonucleotide_primer_pair_laboratory_procedure_performed_name \t\t\t 8\n",
      "i_131_subsequent_administered_dose \t\t\t 8\n",
      "mpnst_neurofibromatosis_heredity \t\t\t 8\n",
      "vascular_tumor_cell_type \t\t\t 7\n",
      "ebv_diagnostic_methodology \t\t\t 7\n",
      "first_recurrent_non_nodal_metastatic_anatomic_site_descriptive_text \t\t\t 7\n",
      "b_lymphocyte_genotyping_method \t\t\t 7\n",
      "histology_of_bone_marrow_sample \t\t\t 6\n",
      "extranocular_nodule_size \t\t\t 6\n",
      "performance_status_assessment_timepoint_category_other_text \t\t\t 6\n",
      "first_diagnosis_age_of_food_allergy \t\t\t 6\n",
      "lymph_node_location_positive_pathology_text \t\t\t 6\n",
      "days_to_well_differentiated_liposarcoma_resection \t\t\t 6\n",
      "ss18_ssx_fusion_status \t\t\t 6\n",
      "metastatic_breast_carcinoma_estrogen_receptor_level_cell_percent_category \t\t\t 6\n",
      "bilateral_diagnosis_timing_type \t\t\t 5\n",
      "prior_infectious_disease \t\t\t 5\n",
      "days_to_bilateral_tumor_dx \t\t\t 5\n",
      "post_orchi_testosterone \t\t\t 5\n",
      "metastatic_neoplasm_confirmed_diagnosis_method_text \t\t\t 5\n",
      "metastatic_breast_carcinoma_erbb2_immunohistochemistry_level_result \t\t\t 5\n",
      "resp_maint_from_bcg_admin_month_dur \t\t\t 5\n",
      "days_to_well_differentiated_liposarcoma_primary_dx \t\t\t 5\n",
      "brachytherapy_method_other_specify_text \t\t\t 5\n",
      "metastatic_breast_carcinoma_progesterone_receptor_level_cell_percent_category \t\t\t 4\n",
      "inherited_genetic_syndrome_result \t\t\t 4\n",
      "extranodal_disease_involvement_site_other \t\t\t 4\n",
      "metastatic_site \t\t\t 4\n",
      "her2_and_centromere_17_positive_finding_other_measurement_scale_text \t\t\t 4\n",
      "prior_immunologic_disease_other \t\t\t 4\n",
      "metastatic_breast_carcinoma_her2_erbb_pos_finding_cell_percent_category \t\t\t 3\n",
      "undescended_testis_method_right \t\t\t 3\n",
      "radiation_therapy_administered_dose_text \t\t\t 3\n",
      "cholangitis_tissue_evidence \t\t\t 3\n",
      "prior_immunosuppressive_therapy_type \t\t\t 3\n",
      "testis_tumor_macroextent_other \t\t\t 3\n",
      "human_papillomavirus_laboratory_procedure_performed_text \t\t\t 3\n",
      "other_genotyping_outcome_lab_results_text \t\t\t 2\n",
      "surgical_procedure_name \t\t\t 2\n",
      "patient_history_immune_system_and_related_disorders_text \t\t\t 2\n",
      "radiation_therapy_not_administered_specify \t\t\t 2\n",
      "radiation_type_notes \t\t\t 2\n",
      "metastatic_breast_carcinoma_her2_neu_chromosone_17_signal_ratio_value \t\t\t 2\n",
      "prior_immunosuppressive_therapy_other \t\t\t 2\n",
      "percent_tumor_sarcomatoid \t\t\t 2\n",
      "sdhb \t\t\t 1\n",
      "pre_orchi_testosterone \t\t\t 1\n",
      "sdha \t\t\t 1\n",
      "sdhaf2_sdh5 \t\t\t 1\n",
      "vhl \t\t\t 1\n",
      "pos_finding_metastatic_breast_carcinoma_estrogen_receptor_other_measuremenet_scale_text \t\t\t 1\n",
      "other_metastatic_site \t\t\t 1\n",
      "post_orchi_lh \t\t\t 1\n",
      "synchronous_tumor_histology_pct \t\t\t 1\n",
      "hist_hepato_carcinoma_risk \t\t\t 1\n",
      "synchronous_tumor_histology_type \t\t\t 1\n",
      "ret \t\t\t 1\n",
      "metastatic_breast_carcinoma_immunohistochemistry_pr_pos_cell_score \t\t\t 1\n",
      "mpnst_specific_mutations \t\t\t 1\n",
      "percent_ebv_positive_malignant_cells \t\t\t 1\n",
      "pre_orchi_lh \t\t\t 1\n",
      "sdhc \t\t\t 1\n",
      "sdhd \t\t\t 1\n",
      "chemotherapy_negation_radiation_therapy_concurrent_administered_text \t\t\t 1\n",
      "tmem127 \t\t\t 1\n",
      "metastatic_breast_carcinoma_immunohistochemistry_er_pos_cell_score \t\t\t 1\n",
      "mib1_positive_percentage_range.1 \t\t\t 0\n",
      "days_to_first_complete_response \t\t\t 0\n",
      "lymph_node_involvement_site \t\t\t 0\n",
      "days_to_first_partial_response \t\t\t 0\n",
      "epstein_barr_viral_status \t\t\t 0\n",
      "genetic_abnormality_method_other \t\t\t 0\n",
      "eml4_alk_translocation_result \t\t\t 0\n",
      "genetic_abnormality_results_other \t\t\t 0\n",
      "days_to_first_response \t\t\t 0\n",
      "genetic_abnormality_tested_other \t\t\t 0\n",
      "genetic_abnormality_tested.1 \t\t\t 0\n",
      "genetic_abnormality_results \t\t\t 0\n",
      "pos_lymph_node_location \t\t\t 0\n",
      "pet_scan_results \t\t\t 0\n",
      "history_of_other_malignancy \t\t\t 0\n",
      "extranodal_involvment_site_other \t\t\t 0\n",
      "days_to_patient_progression_free \t\t\t 0\n",
      "live_birth_number \t\t\t 0\n",
      "lost_follow_up \t\t\t 0\n",
      "history_immunological_disease \t\t\t 0\n",
      "hysterectomy_Performed_Ind-3 \t\t\t 0\n",
      "laboratory_procedure_tumor_marker_squamous_cell_carcinoma_antigen_result_value \t\t\t 0\n",
      "patient_pregnancy_count \t\t\t 0\n",
      "prescribed_dose_units \t\t\t 0\n",
      "days_to_laboratory_procedure_tumor_marker_squamous_cell_carcinoma_antigen_result \t\t\t 0\n",
      "history_immunosuppressive_dx_other \t\t\t 0\n",
      "oligonucleotide_primer_pair_laboratory_procedure_performed_text \t\t\t 0\n",
      "stage_other \t\t\t 0\n",
      "history_of_radiation_metastatic_site \t\t\t 0\n",
      "eml4_alk_translocation_identified \t\t\t 0\n",
      "hpv_test \t\t\t 0\n",
      "bone_marrow_sample_histology \t\t\t 0\n",
      "genetic_abnormality_method \t\t\t 0\n",
      "history_immunological_disease_other \t\t\t 0\n",
      "ldh_norm_range_upper \t\t\t 0\n",
      "ldh_level \t\t\t 0\n",
      "egfr_mutation_identified \t\t\t 0\n",
      "er_estimated_duration_response \t\t\t 0\n",
      "pos_lymph_node_location_other \t\t\t 0\n",
      "metastatic_breast_carcinoma_estrogen_receptor_detection_method_text \t\t\t 0\n",
      "patient_progression_status \t\t\t 0\n",
      "history_prior_surgery_type_other \t\t\t 0\n",
      "pet_ct_standardized_values \t\t\t 0\n",
      "history_prior_surgery_indicator \t\t\t 0\n",
      "field \t\t\t 0\n",
      "molecular_abnormality_results_other \t\t\t 0\n",
      "molecular_abnormality_results \t\t\t 0\n",
      "metastatic_breast_carcinoma_pos_finding_other_scale_measurement_text \t\t\t 0\n",
      "microvascular_density_mvd \t\t\t 0\n",
      "death_cause_text \t\t\t 0\n",
      "nadir_cd4_counts \t\t\t 0\n",
      "days_to_sample_procurement \t\t\t 0\n",
      "hbv_test \t\t\t 0\n",
      "on_haart_therapy_at_cancer_diagnosis \t\t\t 0\n",
      "history_immunosuppresive_dx \t\t\t 0\n",
      "hcv_test \t\t\t 0\n",
      "prior_aids_conditions \t\t\t 0\n",
      "on_haart_therapy_prior_to_cancer_diagnosis \t\t\t 0\n",
      "metastatic_breast_carcinoma_her2_erbb_method_calculation_method_text \t\t\t 0\n",
      "kshv_hhv8_test \t\t\t 0\n",
      "days_to_hiv_diagnosis \t\t\t 0\n",
      "metastatic_breast_carcinoma_her2_erbb_pos_finding_fluorescence_in_situ_hybridization_calculation_method_text \t\t\t 0\n",
      "hiv_status \t\t\t 0\n",
      "hiv_rna_load_at_diagnosis \t\t\t 0\n",
      "cdc_hiv_risk_group \t\t\t 0\n",
      "history_prior_surgery_type \t\t\t 0\n",
      "history_of_radiation_primary_site \t\t\t 0\n",
      "days_to_tumor_progression \t\t\t 0\n",
      "er_response_type \t\t\t 0\n",
      "metastatic_breast_carcinoma_progesterone_receptor_detection_method_text \t\t\t 0\n",
      "follicular_component_percent \t\t\t 0\n",
      "metastatic_breast_carcinoma_pos_finding_progesterone_receptor_other_measure_scale_text \t\t\t 0\n",
      "tumor_resected_max_dimension \t\t\t 0\n",
      "ebv_status_malignant_cells_method \t\t\t 0\n",
      "immunophenotypic_analysis_tested \t\t\t 0\n",
      "extranodal_sites_involvement_number \t\t\t 0\n",
      "er_disease_extent_prior_er_treatment \t\t\t 0\n",
      "maximum_tumor_bulk_anatomic_site \t\t\t 0\n",
      "immunophenotypic_analysis_results \t\t\t 0\n",
      "history_relevant_infectious_dx \t\t\t 0\n",
      "her2_neu_metastatic_breast_carcinoma_copy_analysis_input_total_number \t\t\t 0\n",
      "igh_genotype_results.1 \t\t\t 0\n",
      "her2_neu_and_centromere_17_copy_number_metastatic_breast_carcinoma_analysis_input_total_number_count \t\t\t 0\n",
      "b_lymphocyte_genotyping_method.1 \t\t\t 0\n",
      "ebv_positive_malignant_cells_percent \t\t\t 0\n",
      "prior_infectious_disease_other \t\t\t 0\n",
      "bone_marrow_involvement.1 \t\t\t 0\n",
      "metastatic_breast_carcinoma_fluorescence_in_situ_hybridization_diagnostic_proc_centromere_17_signal_result_range \t\t\t 0\n",
      "history_relevant_infectious_dx_other \t\t\t 0\n",
      "metastatic_breast_carcinoma_pos_finding_her2_erbb2_other_measure_scale_text \t\t\t 0\n",
      "immunophenotypic_analysis_method \t\t\t 0\n",
      "igk_genotype_results.1 \t\t\t 0\n",
      "er_solid_tumor_response_documented_type \t\t\t 0\n",
      "er_solid_tumor_response_documented_type_other \t\t\t 0\n",
      "cd4_counts_at_diagnosis \t\t\t 0\n"
     ]
    }
   ],
   "source": [
    "fu = pd.read_csv('clinical_PANCAN_patient_with_followup.tsv',sep='\\t', encoding='latin1')\n",
    "nvalues = []\n",
    "for i in fu.columns: \n",
    "    fu_drop = fu[i].dropna()\n",
    "    fu_drop = fu_drop[fu_drop!='[Not Available]']\n",
    "    fu_drop = fu_drop[fu_drop!='[Not Applicable]']\n",
    "    nvalues.append([i,len(fu_drop)])\n",
    "nvalues = pd.DataFrame(nvalues,columns=['column','nvalues'])\n",
    "nvalues_sorted = nvalues.sort_values('nvalues',ascending=False)\n",
    "for i in nvalues_sorted.iterrows():\n",
    "    print(i[1]['column'],'\\t\\t\\t',i[1]['nvalues'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nunique 169\n",
      "['C74.0' 'C67.4' 'C67.2' 'C67.9' 'C67.0' 'C67.1' 'C67.3' 'C67.5' 'C67.6'\n",
      " 'C50.9' 'C50.919' 'C50.5' 'C50.2' 'C50.3' 'C50.8' 'C50.4' 'C53.9' 'C53.0'\n",
      " 'C53.1' 'C56.9' '[Not Available]' 'C22.1' 'C24.0' 'C22.0' 'C18.0' 'C18.2'\n",
      " 'C18.6' 'C18.7' 'C18.9' 'C18.4' 'C18.3' 'C18.5' 'C19' 'C83.3' 'C48.1'\n",
      " 'C38.1' 'C08.0' 'C49.0' 'C77.1' 'C62.90' 'C71.7' 'C71.6' 'C15.5' 'C15.9'\n",
      " 'C15.4' 'C15.3' 'C16.0' 'C71.9' 'C71.4' 'C71.1' 'C71.8' 'C71.2' 'C02.9'\n",
      " 'C32.9' 'C04.9' 'C06.0' 'C14.8' 'C09.9' 'C01' 'C10.3' 'C10.9' 'C32.1'\n",
      " 'C03.9' 'C13.9' 'C06.9' 'C05.0' 'C05.9' 'C04.0' 'C02.2' 'C41.1' 'C02.1'\n",
      " 'C00.9' 'C06.2' 'C03.0' 'C03.1' 'C64.9' 'C64.1' 'C71.0' 'C71.3' 'C34.3'\n",
      " 'C34.1' 'C34.2' 'C34.9' 'C34.30' 'C34.0' 'C34.8' 'C34.12' 'C34.10'\n",
      " 'C45.0' 'C34.90' 'C38.4' 'C48.0' 'C48.2' 'C25.0' 'C25.8' 'C25.1' 'C25.9'\n",
      " 'C25.2' 'C74.9' 'C74.1' 'C49.3' 'C76.1' 'C49.6' 'C38.0' 'C76.0' 'C49.4'\n",
      " 'C49.5' 'C75.5' 'C72.9' 'C38.3' 'C61' 'C61.9' 'C20' 'C80.1' 'C49.1'\n",
      " 'C49.2' 'C55' 'C63.1' 'C40.3' 'C49.9' 'C49.8' 'C49.10' 'C16.9' 'C54.9'\n",
      " 'C40.2' 'C54.2' 'C54.0' 'C70.1' 'C47.1' 'C44.5' 'C77.3' 'C77.0' 'C44.9'\n",
      " 'C44.601' 'C44.50' 'C44.701' 'C77.9' 'C77.4' 'C77.5' 'C44.6' 'C17.9'\n",
      " 'C44.31' 'C44.3' 'C44.7' 'C44.4' 'C51.9' 'C43.51' 'C44.2' 'C76.2' 'C52'\n",
      " 'C07' 'C76.3' 'C72.0' 'C41.0' 'C77.2' 'C49.20' 'C54.1' 'C16.5' 'C16.2'\n",
      " 'C16.1' 'C16.3' 'C62.9' 'C73' 'C37' 'C54.3' 'C69.80' 'C69.30' 'C69.40'\n",
      " 'C69.90']\n",
      "icd_10\n",
      "C50.9     1090\n",
      "C64.9      652\n",
      "C71.9      652\n",
      "C56.9      582\n",
      "C34.1      577\n",
      "          ... \n",
      "C05.9        1\n",
      "C44.50       1\n",
      "C10.3        1\n",
      "C71.8        1\n",
      "C69.90       1\n",
      "Name: count, Length: 169, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "inspect(fu,'icd_10')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['TCGA-CDR', 'TCGA-CDR_Notes', 'ExtraEndpoints', 'ExtraEndpoints_Notes', 'Table4_PHAssumptionTests', 'Table5_PHAssumptionTests', 'TSS_Info', 'Fig2EFG_AdditionalInfo'])"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdr = pd.read_excel('TCGA-CDR-SupplementalTableS1.xlsx', sheet_name = None)\n",
    "cdr.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'bcr_patient_barcode', 'type',\n",
       "       'age_at_initial_pathologic_diagnosis', 'gender', 'race',\n",
       "       'ajcc_pathologic_tumor_stage', 'clinical_stage', 'histological_type',\n",
       "       'histological_grade', 'initial_pathologic_dx_year', 'menopause_status',\n",
       "       'birth_days_to', 'vital_status', 'tumor_status', 'last_contact_days_to',\n",
       "       'death_days_to', 'cause_of_death', 'new_tumor_event_type',\n",
       "       'new_tumor_event_site', 'new_tumor_event_site_other',\n",
       "       'new_tumor_event_dx_days_to', 'treatment_outcome_first_course',\n",
       "       'margin_status', 'residual_tumor', 'OS', 'OS.time', 'DSS', 'DSS.time',\n",
       "       'DFI', 'DFI.time', 'PFI', 'PFI.time', 'Redaction'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdr['TCGA-CDR'].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bcr_patient_barcode</th>\n",
       "      <th>type</th>\n",
       "      <th>PFI.1</th>\n",
       "      <th>PFI.time.1</th>\n",
       "      <th>PFI.2</th>\n",
       "      <th>PFI.time.2</th>\n",
       "      <th>PFS</th>\n",
       "      <th>PFS.time</th>\n",
       "      <th>DSS_cr</th>\n",
       "      <th>DSS.time.cr</th>\n",
       "      <th>DFI.cr</th>\n",
       "      <th>DFI.time.cr</th>\n",
       "      <th>PFI.cr</th>\n",
       "      <th>PFI.time.cr</th>\n",
       "      <th>PFI.1.cr</th>\n",
       "      <th>PFI.time.1.cr</th>\n",
       "      <th>PFI.2.cr</th>\n",
       "      <th>PFI.time.2.cr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TCGA-OR-A5J1</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>754.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TCGA-OR-A5J2</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1677.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>289.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TCGA-OR-A5J3</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2091.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TCGA-OR-A5J4</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>423.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TCGA-OR-A5J5</td>\n",
       "      <td>ACC</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>365.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11155</th>\n",
       "      <td>11156</td>\n",
       "      <td>TCGA-YZ-A980</td>\n",
       "      <td>UVM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1862.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1862.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1862.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1556.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1862.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1862.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11156</th>\n",
       "      <td>11157</td>\n",
       "      <td>TCGA-YZ-A982</td>\n",
       "      <td>UVM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>495.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11157</th>\n",
       "      <td>11158</td>\n",
       "      <td>TCGA-YZ-A983</td>\n",
       "      <td>UVM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>798.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11158</th>\n",
       "      <td>11159</td>\n",
       "      <td>TCGA-YZ-A984</td>\n",
       "      <td>UVM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1396.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1396.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11159</th>\n",
       "      <td>11160</td>\n",
       "      <td>TCGA-YZ-A985</td>\n",
       "      <td>UVM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1184.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11160 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0 bcr_patient_barcode type  PFI.1  PFI.time.1  PFI.2  \\\n",
       "0               1        TCGA-OR-A5J1  ACC    1.0       754.0    1.0   \n",
       "1               2        TCGA-OR-A5J2  ACC    1.0       289.0    1.0   \n",
       "2               3        TCGA-OR-A5J3  ACC    1.0        53.0    1.0   \n",
       "3               4        TCGA-OR-A5J4  ACC    1.0       126.0    1.0   \n",
       "4               5        TCGA-OR-A5J5  ACC    1.0        50.0    1.0   \n",
       "...           ...                 ...  ...    ...         ...    ...   \n",
       "11155       11156        TCGA-YZ-A980  UVM    0.0      1862.0    0.0   \n",
       "11156       11157        TCGA-YZ-A982  UVM    0.0       495.0    0.0   \n",
       "11157       11158        TCGA-YZ-A983  UVM    0.0       798.0    0.0   \n",
       "11158       11159        TCGA-YZ-A984  UVM    0.0      1396.0    0.0   \n",
       "11159       11160        TCGA-YZ-A985  UVM    0.0      1184.0    0.0   \n",
       "\n",
       "       PFI.time.2  PFS  PFS.time  DSS_cr  DSS.time.cr  DFI.cr  DFI.time.cr  \\\n",
       "0           754.0  1.0     754.0     1.0       1355.0     1.0        754.0   \n",
       "1           289.0  1.0     289.0     1.0       1677.0     NaN          NaN   \n",
       "2            53.0  1.0      53.0     0.0       2091.0     1.0         53.0   \n",
       "3           126.0  1.0     126.0     1.0        423.0     NaN          NaN   \n",
       "4            50.0  1.0      50.0     1.0        365.0     NaN          NaN   \n",
       "...           ...  ...       ...     ...          ...     ...          ...   \n",
       "11155      1862.0  1.0    1556.0     0.0       1862.0     NaN          NaN   \n",
       "11156       495.0  0.0     495.0     0.0        495.0     NaN          NaN   \n",
       "11157       798.0  0.0     798.0     0.0        798.0     NaN          NaN   \n",
       "11158      1396.0  1.0     154.0     1.0       1396.0     NaN          NaN   \n",
       "11159      1184.0  0.0    1184.0     0.0       1184.0     NaN          NaN   \n",
       "\n",
       "       PFI.cr  PFI.time.cr  PFI.1.cr  PFI.time.1.cr  PFI.2.cr  PFI.time.2.cr  \n",
       "0         1.0        754.0       1.0          754.0       1.0          754.0  \n",
       "1         1.0        289.0       1.0          289.0       1.0          289.0  \n",
       "2         1.0         53.0       1.0           53.0       1.0           53.0  \n",
       "3         1.0        126.0       1.0          126.0       1.0          126.0  \n",
       "4         1.0         50.0       1.0           50.0       1.0           50.0  \n",
       "...       ...          ...       ...            ...       ...            ...  \n",
       "11155     1.0       1556.0       0.0         1862.0       0.0         1862.0  \n",
       "11156     0.0        495.0       0.0          495.0       0.0          495.0  \n",
       "11157     0.0        798.0       0.0          798.0       0.0          798.0  \n",
       "11158     1.0        154.0       2.0         1396.0       2.0         1396.0  \n",
       "11159     0.0       1184.0       0.0         1184.0       0.0         1184.0  \n",
       "\n",
       "[11160 rows x 19 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdr['ExtraEndpoints']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nunique 78\n",
      "[58. 44. 23. 30. 29. 66. 22. 53. 52. 37. 57. 17. 69. 61. 32. 65. 49. 36.\n",
      " 25. 26. 40. 45. 55. 47. 50. 68. 60. 48. 64. 59. 56. 39. 20. 72. 42. 83.\n",
      " 67. 77. 71. 14. 74. 46. 54. 62. 75. 31. 34. 51. 27. 63. 79. 73. 80. 76.\n",
      " 70. 81. 78. 82. 41. 85. 84. 87. 86. 43. 90. 89. 88. 35. 38. 28. 33. nan\n",
      " 21. 24. 18. 10. 19. 16. 15.]\n",
      "age_at_initial_pathologic_diagnosis\n",
      "60.0    340\n",
      "63.0    330\n",
      "61.0    327\n",
      "58.0    308\n",
      "62.0    308\n",
      "       ... \n",
      "17.0      9\n",
      "14.0      5\n",
      "15.0      3\n",
      "16.0      2\n",
      "10.0      1\n",
      "Name: count, Length: 78, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "inspect(cdr['TCGA-CDR'],'age_at_initial_pathologic_diagnosis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepV_a100",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
