{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dandreas/.conda/envs/deepV_a100/lib/python3.10/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading data from ../aa/tumors.pkl\n",
      "loading data from ../data_processing/consolidated_data.pkl\n",
      "loading data from ../data_processing/dna_seq_by_hgncId.pkl\n",
      "loading data from ../aa/canonical_ref.pkl\n",
      "loading data from ../aa/canonical_mut.pkl\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import gzip\n",
    "import numpy as np\n",
    "from saveAndLoad import *\n",
    "\n",
    "tumors = pickleLoad('../aa/tumors.pkl')\n",
    "data = pickleLoad('../data_processing/consolidated_data.pkl')\n",
    "seqs = pickleLoad('../data_processing/dna_seq_by_hgncId.pkl')\n",
    "ref_aa = pickleLoad('../aa/canonical_ref.pkl')\n",
    "mut_aa = pickleLoad('../aa/canonical_mut.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enumerate_150_mutation_types():\n",
    "    \"\"\"\n",
    "    Return a list of all 150 mutation types:\n",
    "      - 6 single-base changes\n",
    "      - 48 double-base contexts\n",
    "      - 96 triple-base contexts\n",
    "    in strings like:\n",
    "      'T>C'       # single\n",
    "      'AT>C'      # double, 5' base\n",
    "      'T>CG'      # double, 3' base\n",
    "      'AT>CG'     # triple (5' + mutation + 3')\n",
    "    \"\"\"\n",
    "\n",
    "    # We assume the 'pyrimidine-based' set of ref->alt:\n",
    "    #   C->A, C->G, C->T, T->A, T->C, T->G\n",
    "    single_changes = [\n",
    "        (\"C\", \"A\"),\n",
    "        (\"C\", \"G\"),\n",
    "        (\"C\", \"T\"),\n",
    "        (\"T\", \"A\"),\n",
    "        (\"T\", \"C\"),\n",
    "        (\"T\", \"G\"),\n",
    "    ]\n",
    "    # Possible flanking bases\n",
    "    neighbors = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "    # 1) Enumerate 6 single-base changes.\n",
    "    singles = []\n",
    "    for (ref, alt) in single_changes:\n",
    "        singles.append(f\"{ref}>{alt}\")\n",
    "\n",
    "    # 2) Enumerate 48 doubles:\n",
    "    #    We split them into 24 \"5' base + mutation\" and\n",
    "    #    24 \"mutation + 3' base\".\n",
    "    doubles = []\n",
    "\n",
    "    # 2a) 5′ neighbor + mutation, e.g. \"AT>C\"\n",
    "    for five_prime in neighbors:\n",
    "        for (ref, alt) in single_changes:\n",
    "            doubles.append(f\"{five_prime}{ref}>{alt}\")\n",
    "\n",
    "    # 2b) mutation + 3′ neighbor, e.g. \"T>CG\"\n",
    "    for (ref, alt) in single_changes:\n",
    "        for three_prime in neighbors:\n",
    "            doubles.append(f\"{ref}>{alt}{three_prime}\")\n",
    "\n",
    "    # 3) Enumerate 96 triples: \"5' base + mutation + 3' base\"\n",
    "    #    e.g. \"AT>CG\" means five_prime=\"A\", (ref=\"T\", alt=\"C\"), three_prime=\"G\".\n",
    "    triples = []\n",
    "    for five_prime in neighbors:\n",
    "        for (ref, alt) in single_changes:\n",
    "            for three_prime in neighbors:\n",
    "                triples.append(f\"{five_prime}{ref}>{alt}{three_prime}\")\n",
    "\n",
    "    # Combine them: 6 + 48 + 96 = 150 total\n",
    "    all_150 = singles + doubles + triples\n",
    "    return all_150\n",
    "\n",
    "def to_snv150(tumors, barcode, data, seqs, ref_aa, mut_aa, save=False):\n",
    "    snv150 = enumerate_150_mutation_types()\n",
    "    snv150_indexes = {snv:i for i,snv in enumerate(snv150)}\n",
    "    instances = []\n",
    "    counts = np.array([0]*150)\n",
    "    comp = {'A':'T','T':'A','C':'G','G':'C'}\n",
    "\n",
    "    # Grab the indexes from the tumors structure\n",
    "    data_idxs = [i[-1] for i in tumors[barcode]]\n",
    "    ref_idxs = [i[4] for i in tumors[barcode]]\n",
    "    mut_idxs = [i[5] for i in tumors[barcode]]\n",
    "\n",
    "\n",
    "    for data_idx, ref_idx, mut_idx in zip(data_idxs, ref_idxs, mut_idxs):\n",
    "        start,end,chrom,build,variant_class,isoforms,hgncId,gene_start,gene_end,ref_allele,strand,mutation,bc = data[data_idx]\n",
    "\n",
    "        # Pull reference and mutant amino acids\n",
    "        ref_seq, mut_seq = ref_aa[ref_idx], mut_aa[mut_idx]\n",
    "        if (ref_seq is None) or (mut_seq is None):\n",
    "            continue\n",
    "        if '*' in ref_seq: \n",
    "            continue\n",
    "\n",
    "        # Prepare the position, ref, alt, etc.\n",
    "        pos = start\n",
    "        ref = ref_allele\n",
    "        alt = mutation\n",
    "        \n",
    "        # Retrieve the entire reference DNA sequence for this gene\n",
    "        gene_dna_seq = seqs[(build, chrom)][hgncId]\n",
    "\n",
    "        # Sanity-check that the base in gene_dna_seq matches ref_allele\n",
    "        if (len(ref) == 1) and (len(alt) == 1):\n",
    "            r = pos - gene_start\n",
    "            five,ref,three = gene_dna_seq[r-1:r+2]\n",
    "            assert ref == ref_allele, (five,ref,three)\n",
    "        else: continue\n",
    "\n",
    "        if ref in 'AG':\n",
    "            five,ref,alt,three = tuple(map(comp.get, [five,ref,alt,three]))\n",
    "\n",
    "        triple = f'{five}{ref}>{alt}{three}'\n",
    "        single = triple[1:4]\n",
    "        double_5 = triple[:4]\n",
    "        double_3 = triple[1:]\n",
    "\n",
    "        instances+=[single, double_5, double_3, triple]\n",
    "    \n",
    "    for i in instances:\n",
    "        counts[snv150_indexes[i]]+=1\n",
    "    \n",
    "    return counts\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 143530/143530 [00:06<00:00, 21089.30it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "class_data = pd.read_csv('../labeled_data/data_1_00percentMinCancerType.csv')\n",
    "barcodes = class_data['barcode'].values\n",
    "\n",
    "snv150_data = {}\n",
    "for barcode in tqdm(barcodes):\n",
    "    mut_type_vector = to_snv150(tumors, barcode, data, seqs, ref_aa, mut_aa, save=True)\n",
    "    snv150_data[barcode] = mut_type_vector\n",
    "\n",
    "snv150_df = pd.DataFrame(snv150_data).T\n",
    "snv150_df.columns = enumerate_150_mutation_types()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>AC&gt;A</th>\n",
       "      <th>AC&gt;G</th>\n",
       "      <th>AC&gt;T</th>\n",
       "      <th>AT&gt;A</th>\n",
       "      <th>...</th>\n",
       "      <th>TT&gt;AG</th>\n",
       "      <th>TT&gt;AT</th>\n",
       "      <th>TT&gt;CA</th>\n",
       "      <th>TT&gt;CC</th>\n",
       "      <th>TT&gt;CG</th>\n",
       "      <th>TT&gt;CT</th>\n",
       "      <th>TT&gt;GA</th>\n",
       "      <th>TT&gt;GC</th>\n",
       "      <th>TT&gt;GG</th>\n",
       "      <th>TT&gt;GT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GENIE-JHU-00006-00185</th>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-JHU-00007-00187</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-JHU-00024-00426</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-JHU-00026-01223</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-JHU-00032-00469</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-PROV-358b6384eb-e74ad60cbf</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-PROV-a09878f384-bb63e20e8d</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-PROV-dab2e48af5-35f92dc0e9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-PROV-95b5a6d4e0-2ee7ad86ea</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENIE-PROV-9c7ad58b71-a41e571979</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>143530 rows × 150 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  C>A  C>G  C>T  T>A  T>C  T>G  AC>A  AC>G  \\\n",
       "GENIE-JHU-00006-00185               5    1    5    1    5    0     1     0   \n",
       "GENIE-JHU-00007-00187               0    0    1    0    0    0     0     0   \n",
       "GENIE-JHU-00024-00426               2    1    1    0    0    0     1     0   \n",
       "GENIE-JHU-00026-01223               1    0    0    1    0    0     1     0   \n",
       "GENIE-JHU-00032-00469               0    0    2    0    0    0     0     0   \n",
       "...                               ...  ...  ...  ...  ...  ...   ...   ...   \n",
       "GENIE-PROV-358b6384eb-e74ad60cbf    0    0    5    0    1    0     0     0   \n",
       "GENIE-PROV-a09878f384-bb63e20e8d    0    1    0    0    0    0     0     0   \n",
       "GENIE-PROV-dab2e48af5-35f92dc0e9    0    0    1    1    0    1     0     0   \n",
       "GENIE-PROV-95b5a6d4e0-2ee7ad86ea    0    0    0    1    0    0     0     0   \n",
       "GENIE-PROV-9c7ad58b71-a41e571979    0    0    1    0    0    0     0     0   \n",
       "\n",
       "                                  AC>T  AT>A  ...  TT>AG  TT>AT  TT>CA  TT>CC  \\\n",
       "GENIE-JHU-00006-00185                2     0  ...      0      0      0      0   \n",
       "GENIE-JHU-00007-00187                0     0  ...      0      0      0      0   \n",
       "GENIE-JHU-00024-00426                0     0  ...      0      0      0      0   \n",
       "GENIE-JHU-00026-01223                0     1  ...      0      0      0      0   \n",
       "GENIE-JHU-00032-00469                1     0  ...      0      0      0      0   \n",
       "...                                ...   ...  ...    ...    ...    ...    ...   \n",
       "GENIE-PROV-358b6384eb-e74ad60cbf     0     0  ...      0      0      0      0   \n",
       "GENIE-PROV-a09878f384-bb63e20e8d     0     0  ...      0      0      0      0   \n",
       "GENIE-PROV-dab2e48af5-35f92dc0e9     0     0  ...      0      0      0      0   \n",
       "GENIE-PROV-95b5a6d4e0-2ee7ad86ea     0     0  ...      0      0      0      0   \n",
       "GENIE-PROV-9c7ad58b71-a41e571979     0     0  ...      0      0      0      0   \n",
       "\n",
       "                                  TT>CG  TT>CT  TT>GA  TT>GC  TT>GG  TT>GT  \n",
       "GENIE-JHU-00006-00185                 1      1      0      0      0      0  \n",
       "GENIE-JHU-00007-00187                 0      0      0      0      0      0  \n",
       "GENIE-JHU-00024-00426                 0      0      0      0      0      0  \n",
       "GENIE-JHU-00026-01223                 0      0      0      0      0      0  \n",
       "GENIE-JHU-00032-00469                 0      0      0      0      0      0  \n",
       "...                                 ...    ...    ...    ...    ...    ...  \n",
       "GENIE-PROV-358b6384eb-e74ad60cbf      0      0      0      0      0      0  \n",
       "GENIE-PROV-a09878f384-bb63e20e8d      0      0      0      0      0      0  \n",
       "GENIE-PROV-dab2e48af5-35f92dc0e9      0      0      0      0      0      0  \n",
       "GENIE-PROV-95b5a6d4e0-2ee7ad86ea      0      0      0      0      0      0  \n",
       "GENIE-PROV-9c7ad58b71-a41e571979      0      0      0      0      0      0  \n",
       "\n",
       "[143530 rows x 150 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv150_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepV_a100",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
