{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from saveAndLoad import *\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from tqdm import tqdm\n",
    "import concurrent.futures\n",
    "\n",
    "def printAllHead(df):\n",
    "    with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.max_colwidth', None):\n",
    "        print(df.head())\n",
    "def printAll(x):\n",
    "    with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'display.max_colwidth', None):\n",
    "        print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load all data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pickleLoad('consolidated_data.pkl')\n",
    "seqs = pickleLoad('dna_seq_by_hgncId.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class orderedSet:\n",
    "    def __init__(self,item_is_dict=False):\n",
    "        self.ordered_set = dict()\n",
    "        self.counter = 0\n",
    "        self.item_is_dict = item_is_dict\n",
    "    \n",
    "    def add_to_set(self,item):\n",
    "        if self.item_is_dict:\n",
    "            item = tuple(sorted(item.items()))\n",
    "        if item not in self.ordered_set:\n",
    "            self.ordered_set[item] = self.counter\n",
    "            self.counter+=1\n",
    "            return self.counter -1\n",
    "        return self.get_index(item)\n",
    "    \n",
    "    def get_index(self,item):\n",
    "        if type(item)==dict:\n",
    "            item = tuple(sorted(item.items()))\n",
    "        return self.ordered_set.get(item,None)\n",
    "    \n",
    "    def return_ordered_list(self):\n",
    "        if self.item_is_dict:\n",
    "            return [dict(i) for i in list(self.ordered_set.keys())]\n",
    "        return list(self.ordered_set.keys())\n",
    "\n",
    "class mutatedSequenceGenerator:\n",
    "    def __init__(self,data,seqs):\n",
    "        self.data = data\n",
    "        self.seqs = seqs\n",
    "    \n",
    "    def printInfo(self,idx):\n",
    "        start,end,chrom,build,variant_class,isoforms,hgncId,gene_start,gene_end,ref_allele,strand,mutation_value,barcode = self.data[idx]\n",
    "        print('variant class:',variant_class)\n",
    "        print('mutation start:',start)\n",
    "        print('mutation end:',end)\n",
    "        print('chromosome',chrom)\n",
    "        print('build',build)\n",
    "        print('hgnc id:',hgncId)\n",
    "        print('gene start:',gene_start)\n",
    "        print('gene end:',gene_end)\n",
    "        print('isoforms:',isoforms)\n",
    "        print('reference allele:',ref_allele)\n",
    "        print('strand:',strand)\n",
    "        print('mutation:',mutation_value)\n",
    "        print('tumor sample barcode:',barcode)\n",
    "    \n",
    "    def rc(self,s): #reverse complement\n",
    "        reverse_complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A','N':'N'}\n",
    "        return ''.join([reverse_complement[i] for i in s[::-1]])\n",
    "    \n",
    "    def mutate(self,mutation,mut_start_zero,mut_end_zero,exon_start_zero,exon_end_zero,variant_class,ref_gene,ref_allele):\n",
    "        \n",
    "        if 'Ins' in variant_class:\n",
    "            mutated_gene = ref_gene[:mut_start_zero-1] + mutation + ref_gene[mut_start_zero-1:]\n",
    "            mutated_exon = mutated_gene[exon_start_zero-1:exon_end_zero + len(mutation)]\n",
    "\n",
    "        elif 'Del' in variant_class:\n",
    "            mutated_gene = ref_gene[:mut_start_zero-1] + mutation + ref_gene[mut_end_zero:]\n",
    "            mutated_exon = mutated_gene[exon_start_zero-1:exon_end_zero - len(ref_allele)]\n",
    "\n",
    "        else:\n",
    "            mutated_gene = ref_gene[:mut_start_zero-1] + mutation + ref_gene[mut_end_zero:]\n",
    "            mutated_exon = mutated_gene[exon_start_zero-1:exon_end_zero]\n",
    "\n",
    "        return mutated_exon\n",
    "\n",
    "    def getRefAndMutatedSeq(self,idx,verbose=False):\n",
    "        start,end,chrom,build,variant_class,isoforms,hgncId,gene_start,gene_end,ref_allele,strand,mutation,barcode = self.data[idx]\n",
    "\n",
    "        # ref_gene = ref_genome[gene_start-1:gene_end]\n",
    "        ref_gene = self.seqs[(build,chrom)][hgncId]\n",
    "\n",
    "        reverse = strand=='-'\n",
    "\n",
    "        # mutation = self.getMutation(tumorSeqAllele1,tumorSeqAllele2,ref_allele,variant_class)\n",
    "\n",
    "        start_zero = start - gene_start + 1\n",
    "        end_zero = end - gene_start + 1\n",
    "        mutated_gene = self.mutate(mutation,start_zero,end_zero,1,gene_end - gene_start + 1,variant_class,ref_gene,ref_allele)\n",
    "\n",
    "        #isoforms mutated\n",
    "        isoform_mutated_seqs = {}\n",
    "        isoform_seqs = {}\n",
    "        all_exons_seq = ''\n",
    "        all_exons_mutated_seq = ''\n",
    "        for idx,isoform in isoforms.items():\n",
    "            isoform_mutated_seqs[idx] = ''\n",
    "            isoform_seqs[idx] = ''\n",
    "            isoform_is_mutated = False\n",
    "            for exon_start,exon_end in isoform:\n",
    "                exon_start_zero = exon_start - gene_start + 1\n",
    "                exon_end_zero = exon_end - gene_start + 1\n",
    "                isoform_seqs[idx]+=ref_gene[exon_start_zero-1:exon_end_zero]\n",
    "                all_exons_seq += ref_gene[exon_start_zero-1:exon_end_zero]\n",
    "                if exon_start <= start and exon_end >= end:\n",
    "                    isoform_is_mutated = True\n",
    "                    start_zero = start - gene_start + 1\n",
    "                    end_zero = end - gene_start + 1\n",
    "                    mutated_exon = self.mutate(mutation,start_zero,end_zero,exon_start_zero,exon_end_zero,variant_class,ref_gene,ref_allele)\n",
    "                    \n",
    "                    if verbose:\n",
    "                        print(ref_gene[exon_start_zero-1:exon_end_zero])\n",
    "                        print(mutated_exon)\n",
    "                        \n",
    "                    isoform_mutated_seqs[idx]+=mutated_exon\n",
    "                    all_exons_mutated_seq += mutated_exon\n",
    "                else: \n",
    "                    isoform_mutated_seqs[idx]+=ref_gene[exon_start_zero-1:exon_end_zero]\n",
    "                    all_exons_mutated_seq += ref_gene[exon_start_zero-1:exon_end_zero]\n",
    "            if not isoform_is_mutated: del isoform_mutated_seqs[idx] \n",
    "        \n",
    "        if verbose:\n",
    "            print(ref_gene[start_zero-20:end_zero+20])\n",
    "            print(mutated_gene[start_zero-20:end_zero+20])\n",
    "\n",
    "        if reverse:\n",
    "            try:\n",
    "                ref_gene = self.rc(ref_gene)\n",
    "            except: print('ref_gene:',ref_gene)\n",
    "            isoform_mutated_seqs = {k:self.rc(v) for k,v in isoform_mutated_seqs.items()}\n",
    "            isoform_seqs = {k:self.rc(v) for k,v in isoform_seqs.items()}\n",
    "            all_exons_seq = self.rc(all_exons_seq)\n",
    "            all_exons_mutated_seq = self.rc(all_exons_mutated_seq)\n",
    "\n",
    "        isoforms_mutated_aa = {k:str(Seq(v).translate()) for k,v in isoform_mutated_seqs.items()}\n",
    "        isoforms_aa = {k:str(Seq(v).translate()) for k,v in isoform_seqs.items()}\n",
    "        all_exons_aa = str(Seq(all_exons_seq).translate())\n",
    "        all_exons_mutated_aa = str(Seq(all_exons_mutated_seq).translate())\n",
    "\n",
    "        # filter synonymous mutations\n",
    "        for k in list(isoforms_mutated_aa.keys()):\n",
    "            iso_aa = isoforms_mutated_aa[k]\n",
    "            ref_aa = isoforms_aa[k]\n",
    "            if iso_aa == ref_aa:\n",
    "                del isoforms_mutated_aa[k]\n",
    "        \n",
    "        if all_exons_aa == all_exons_mutated_aa:\n",
    "            all_exons_mutated_aa = None\n",
    "\n",
    "        # return ref_gene, mutated_gene, isoform_mutated_seqs, isoform_seqs, all_exons_seq, all_exons_mutated_seq, isoforms_mutated_aa, isoforms_aa, all_exons_aa, all_exons_mutated_aa\n",
    "        return isoforms_mutated_aa, isoforms_aa, all_exons_mutated_aa, all_exons_aa, barcode\n",
    "\n",
    "    #create an ordered set of sequences (using dict as ordered set)\n",
    "    #map each mutation idx to the ordered set\n",
    "    def getAllMutations(self):\n",
    "        exons_concat_ref = orderedSet()\n",
    "        exons_concat_mut = orderedSet()\n",
    "        isos_ref = orderedSet(item_is_dict=True)\n",
    "        isos_mut = orderedSet(item_is_dict=True)\n",
    "        canonical_ref = orderedSet()\n",
    "        canonical_mut = orderedSet()\n",
    "\n",
    "        #0:concat_ref\n",
    "        #1:concat_mut\n",
    "        #2:iso_ref\n",
    "        #3:iso_mut\n",
    "        #4:canonical_ref\n",
    "        #5:canonical_mut\n",
    "        tumors = {}\n",
    "\n",
    "        getRefAndMutatedSeq = self.getRefAndMutatedSeq\n",
    "\n",
    "        for idx in tqdm(range(len(self.data))):\n",
    "            isoforms_mutated_aa, isoforms_aa, all_exons_mutated_aa, all_exons_aa, barcode = getRefAndMutatedSeq(idx)\n",
    "\n",
    "            isos_ref_idx = isos_ref.add_to_set(isoforms_aa)\n",
    "            isos_mut_idx = isos_mut.add_to_set(isoforms_mutated_aa)\n",
    "\n",
    "            canonical_ref_aa = isoforms_aa['1'] if '1' in isoforms_aa else None\n",
    "            canonical_mut_aa = isoforms_mutated_aa['1'] if '1' in isoforms_mutated_aa else None\n",
    "            canonical_mut_idx = canonical_mut.add_to_set(canonical_mut_aa)\n",
    "            canonical_ref_idx = canonical_ref.add_to_set(canonical_ref_aa)\n",
    "\n",
    "            exons_concat_ref_idx = exons_concat_ref.add_to_set(all_exons_aa)\n",
    "            exons_concat_mut_idx = exons_concat_mut.add_to_set(all_exons_mutated_aa)\n",
    "\n",
    "            if tumors.get(barcode)is None: tumors[barcode] = []\n",
    "            tumors[barcode].append([exons_concat_ref_idx, exons_concat_mut_idx, isos_ref_idx, isos_mut_idx, canonical_ref_idx, canonical_mut_idx, idx])\n",
    "\n",
    "        ordered_lists = [i.return_ordered_list() for i in [isos_ref,isos_mut,exons_concat_ref,exons_concat_mut, canonical_ref, canonical_mut]]\n",
    "        return ordered_lists, tumors\n",
    "    \n",
    "    # # a list of mutations for every index\n",
    "    # def getAllMutations(self):\n",
    "    #     exons_concat_ref = []\n",
    "    #     exons_concat_mut = []\n",
    "    #     isos_ref = []\n",
    "    #     isos_mut = []\n",
    "    #     canonical_ref = []\n",
    "    #     canonical_mut = []\n",
    "\n",
    "    #     getRefAndMutatedSeq = self.getRefAndMutatedSeq\n",
    "\n",
    "    #     for idx in tqdm(range(len(self.data))):\n",
    "    #         isoforms_mutated_aa, isoforms_aa, all_exons_mutated_aa, all_exons_aa = getRefAndMutatedSeq(idx)\n",
    "    #         isos_ref.append(isoforms_aa)\n",
    "    #         isos_mut.append(isoforms_mutated_aa)\n",
    "    #         canonical_ref_aa = isoforms_aa['1'] if '1' in isoforms_aa else None\n",
    "    #         canonical_mut_aa = isoforms_mutated_aa['1'] if '1' in isoforms_mutated_aa else None\n",
    "    #         canonical_mut.append(canonical_mut_aa)\n",
    "    #         canonical_ref.append(canonical_ref_aa)\n",
    "    #         exons_concat_ref.append(all_exons_aa)\n",
    "    #         exons_concat_mut.append(all_exons_mutated_aa)\n",
    "    #     return isos_ref,isos_mut,exons_concat_ref,exons_concat_mut, canonical_ref, canonical_mut\n",
    "\n",
    "seqGen = mutatedSequenceGenerator(data,seqs)\n",
    "getMutAA_isoforms = lambda x: seqGen.getRefAndMutatedSeq(x)[0]\n",
    "getRefAA_isoforms = lambda x: seqGen.getRefAndMutatedSeq(x)[1]\n",
    "getMutAA_concat = lambda x: seqGen.getRefAndMutatedSeq(x)[2]\n",
    "getRefAA_concat = lambda x: seqGen.getRefAndMutatedSeq(x)[3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# i=0\n",
    "# seqGen.printInfo(i)\n",
    "# isoforms_mutated_aa, isoforms_aa, all_exons_mutated_aa, all_exons_aa = seqGen.getRefAndMutatedSeq(i,verbose=True)\n",
    "# print(isoforms_mutated_aa['1'])\n",
    "# print(isoforms_aa['1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[isos_ref,isos_mut,exons_concat_ref,exons_concat_mut, canonical_ref, canonical_mut], tumors = seqGen.getAllMutations()\n",
    "pickleSave(isos_ref,'../aa/','isos_ref.pkl')\n",
    "pickleSave(isos_mut,'../aa/','isos_mut.pkl')\n",
    "pickleSave(exons_concat_ref,'../aa/','exons_concat_ref.pkl')\n",
    "pickleSave(exons_concat_mut,'../aa/','exons_concat_mut.pkl')\n",
    "pickleSave(canonical_ref,'../aa/','canonical_ref.pkl')\n",
    "pickleSave(canonical_mut,'../aa/','canonical_mut.pkl')\n",
    "pickleSave(tumors,'../aa/','tumors.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## CASE 1\n",
    "\n",
    "# mutation     --> HIST1H3B (previous hugo symbol of H3C2)              \n",
    "# mutation_map --> HGNC:4776              \n",
    "                                        \n",
    "# HGNC:4776 not found in hgnc_uniprot.\n",
    "# look up HGNC:4776 on uniprot --> P68431\n",
    "# look up HGNC:4776 on HGNC --> H3C2, H3 clustered histone 2\n",
    "\n",
    "# uniprot_hgnc['P68431'] --> HGNC:4774\n",
    "\n",
    "# look up HGNC:4774 on HGNC --> H3C12, H3 clustered histone 12\n",
    "# uniprotGtf entry for P68431 contains multiple transcriptIds with _dup1, _dup2\n",
    "# the ENSEMBL ID of dup2 matches H3C2\n",
    "\n",
    "# solution: \n",
    "# 1. create exons for each transcriptId\n",
    "# 2. if there are more than one transcriptId, try to resolve\n",
    "# now these are thrown out, 3148 hgncIds not found in uniprot (0.17%) (probably not all of these are transcriptId issues)\n",
    "\n",
    "## CASE 2\n",
    "\n",
    "# mutation      --> PIK3CA\n",
    "# mutation_map  --> HGNC:8975\n",
    "\n",
    "# hgnc_uniprot              --> C9J951\n",
    "# look up C9J951 on HGNC    --> PIK3CA\n",
    "# look up PIK3CA on uniprot --> P42336\n",
    "\n",
    "# solution:\n",
    "# there are multiple uniprot Ids that map to the same HGNC:ID \n",
    "# write the hgnc_uniprot algorithm to check if there are multiple Ids and prefer the one in SwissProt \n",
    "\n",
    "\n",
    "#### NEED TO CHECK CHROMOSOME ON UNIPROT GTF ####\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for ni,i in enumerate(data):\n",
    "#     if 'Del' in i[4]:\n",
    "#         a1 = type(i[13])==str and len(i[13])>2 \n",
    "#         a2 = type(i[14])==str and len(i[14])>2\n",
    "#         if a1 or a2: assert False, ni\n",
    "#     # if 'Ins' in i[4]:\n",
    "#     #     ar = type(i[12])==str and len(i[12])>2\n",
    "#     #     if ar: assert False, ni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading data from ./aa/isos_mut.pkl\n"
     ]
    }
   ],
   "source": [
    "# from saveAndLoad import *\n",
    "# isos_mut = pickleLoad('./aa/isos_mut.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# asdf = sorted(tuple(list(isos_mut[0].items())))\n",
    "# dict(asdf) == isos_mut[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepV_a100",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
